{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Individual Sequence Length: 3021\n",
      "Number of Experimental Sequences: 5073\n",
      "Number of Chromosomes: 10\n",
      "\n",
      "Sequences per Chromosome: 507\n",
      "Sequences on Last Chromosome: 510\n",
      "Number of sequences per chromosome are properly structured.\n",
      "\n",
      "Bases per Chromosome: 1531647\n",
      "Bases on Last Chromosome: 1540710\n",
      "Total Bases in Experiment: 15325533\n",
      "Number of bases per chromosome are properly structured.\n"
     ]
    }
   ],
   "source": [
    "#toying with seq num and chrom size to deal with edge cases\n",
    "sequence_num = 5073\n",
    "window = 1500\n",
    "chrom_num=10\n",
    "\n",
    "#sequence_num = len(file)\n",
    "seq_length = ((2*int(window))+21)\n",
    "tot_bases = int(sequence_num*seq_length)\n",
    "seq_per_chrom = math.floor(sequence_num/chrom_num)\n",
    "seq_last_chrom = seq_per_chrom + sequence_num%chrom_num\n",
    "#this is to check if there is the correct amount of sequences per chromosome\n",
    "scount = seq_last_chrom + (seq_per_chrom)*(chrom_num-1)\n",
    "\n",
    "base_per_chrom = seq_length*seq_per_chrom \n",
    "base_last_chrom = seq_length*seq_last_chrom\n",
    "#this is to check if there is the correct amount of bases per chromosome\n",
    "bscount = base_last_chrom + (base_per_chrom)*(chrom_num-1)\n",
    "\n",
    "print(\"Individual Sequence Length: \" + str(seq_length))\n",
    "print(\"Number of Experimental Sequences: \" + str(sequence_num)) \n",
    "print(\"Number of Chromosomes: \" + str(chrom_num))\n",
    "print('')\n",
    "print(\"Sequences per Chromosome: \" + str(seq_per_chrom))\n",
    "print(\"Sequences on Last Chromosome: \" + str(seq_last_chrom))\n",
    "if sequence_num-scount != 0:\n",
    "    print('Error! The total number of sequences does not equal the sum of the sequences divided into chromosomes!')\n",
    "    sys.exit(1)\n",
    "else:\n",
    "    print('Number of sequences per chromosome are properly structured.')\n",
    "print('')\n",
    "\n",
    "print(\"Bases per Chromosome: \" + str(base_per_chrom))\n",
    "print(\"Bases on Last Chromosome: \" + str(base_last_chrom))\n",
    "print(\"Total Bases in Experiment: \" + str(tot_bases))\n",
    "if tot_bases-bscount != 0:\n",
    "    print('Error! The total number of bases does not equal the sum of the bases divided into chromosomes!')\n",
    "    sys.exit(1)\n",
    "else:\n",
    "    print('Number of bases per chromosome are properly structured.')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "#funcs needed to run write_fasta\n",
    "def sequence_input(sequence):\n",
    "    position_feq = []\n",
    "\n",
    "    with open(sequence) as seq:\n",
    "\n",
    "        ##remove the header line\n",
    "        lines = seq.readlines()[1:]\n",
    "        for i in range(len(lines)):\n",
    "            pos_prob = []\n",
    "            try:\n",
    "                pos_prob.append(lines[i].strip('\\n').split(',')[1])\n",
    "                pos_prob.append(lines[i].strip('\\n').split(',')[2])\n",
    "                pos_prob.append(lines[i].strip('\\n').split(',')[3])\n",
    "                pos_prob.append(lines[i].strip('\\n').split(',')[4])\n",
    "                position_feq.append(pos_prob)\n",
    "            except IndexError:\n",
    "                \n",
    "                pos_prob.append(lines[i].strip('\\n').split('\\t')[1])\n",
    "                pos_prob.append(lines[i].strip('\\n').split('\\t')[2])\n",
    "                pos_prob.append(lines[i].strip('\\n').split('\\t')[3])\n",
    "                pos_prob.append(lines[i].strip('\\n').split('\\t')[4])\n",
    "                position_feq.append(pos_prob)\n",
    "            else:\n",
    "                print(\"Check the input file. \\n It should be tab or comma separated\")\n",
    "\n",
    "\n",
    "    return(position_feq)\n",
    "\n",
    "def sequence_generator(bases,position_feq, sequence_num):\n",
    "    '''takes in frequencies per position and simulates sequences using a \n",
    "    1st order Markov Model.\n",
    "    '''\n",
    "    \n",
    "    first = [position_feq[0:1]]\n",
    "    last =  [position_feq[-1:]]\n",
    "    print('Position Frequency: Contains ' + str(len(position_feq)) + \" elements. \" + str(first) + '...' + str(last))\n",
    "    \n",
    "    sequences = np.empty([sequence_num, len(position_feq)], dtype=str)\n",
    "    \n",
    "    for i in range(len(position_feq)):\n",
    "        column = np.random.choice(bases, sequence_num, p=position_feq[i])\n",
    "        sequences[:,i] = column\n",
    "        \n",
    "    joined_sequences = [''.join(row) for row in sequences]\n",
    "    \n",
    "    return joined_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----------Reading Per-Base Sequence Frequency----------------\n",
      "Base Content File: /Users/tajo5912/rbg/test/generated_sequences/test_base_content.tsv\n",
      "-----------Generating Sequences-------------------------------\n",
      "Position Frequency: Contains 3001 elements. [[['0.27503343079259307', '0.26821128260918836', '0.23292196459735726', '0.22383332200086126']]]...[[['0.26763899276988284', '0.2770449445842116', '0.22373699598830488', '0.2315790666576007']]]\n"
     ]
    }
   ],
   "source": [
    "#running the scripts to get proper input files for write_fasta\n",
    "outdir = '/Users/tajo5912/rbg/test'\n",
    "sample= 'test'\n",
    "\n",
    "print(\"-----------Reading Per-Base Sequence Frequency----------------\")\n",
    "tsv = outdir + '/generated_sequences/' + sample + '_base_content.tsv'\n",
    "print('Base Content File: ' + tsv)\n",
    "position_prob = sequence_input(tsv)\n",
    "\n",
    "print(\"-----------Generating Sequences-------------------------------\")\n",
    "sequences_generating = sequence_generator(bases=['A', 'T', 'G', 'C'], sequence_num=sequence_num, position_feq=position_prob)\n",
    "\n",
    "# print(\"-----------Writing Sequences to Fasta File--------------------\")\n",
    "# write_fasta(sequences_generating, sample, outdir, int(window), sequence_num, chrom_num)\n",
    "# #print('Simulated Fasta File: ' + outdir + '/generated_sequences/' + str(sample) + \"_simulated.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "#properly formating the output fasta\n",
    "def write_fasta(generated_sequences, sample, outdir, window, sequence_num, chrom_num):\n",
    "    ''' writes sequences generated into fasta file\n",
    "    '''\n",
    "        \n",
    "    #checking fasta parameters\n",
    "    seq_length = ((2*int(window))+21)\n",
    "    tot_bases = int(sequence_num*seq_length)\n",
    "    seq_per_chrom = math.floor(sequence_num/chrom_num)\n",
    "    seq_last_chrom = seq_per_chrom + sequence_num%chrom_num\n",
    "    #this is to check if there is the correct amount of sequences per chromosome\n",
    "    scount = seq_last_chrom + (seq_per_chrom)*(chrom_num-1)\n",
    "\n",
    "    base_per_chrom = seq_length*seq_per_chrom \n",
    "    base_last_chrom = seq_length*seq_last_chrom\n",
    "    #this is to check if there is the correct amount of bases per chromosome\n",
    "    bscount = base_last_chrom + (base_per_chrom)*(chrom_num-1)\n",
    "\n",
    "    print(\"Individual Sequence Length: \" + str(seq_length))\n",
    "    print(\"Number of Experimental Sequences: \" + str(sequence_num)) \n",
    "    print(\"Number of Chromosomes: \" + str(chrom_num) +'\\n')\n",
    "    print(\"Sequences per Chromosome: \" + str(seq_per_chrom))\n",
    "    print(\"Sequences on Last Chromosome: \" + str(seq_last_chrom))\n",
    "    if sequence_num-scount != 0:\n",
    "        print('Error! The total number of sequences does not equal the sum of the sequences divided into chromosomes!')\n",
    "        sys.exit(1)\n",
    "    else:\n",
    "        print('Number of sequences per chromosome are properly structured.\\n')\n",
    "\n",
    "    print(\"Bases per Chromosome: \" + str(base_per_chrom))\n",
    "    print(\"Bases on Last Chromosome: \" + str(base_last_chrom))\n",
    "    print(\"Total Bases in Experiment: \" + str(tot_bases))\n",
    "    if tot_bases-bscount != 0:\n",
    "        print('Error! The total number of bases does not equal the sum of the bases divided into chromosomes!')\n",
    "        sys.exit(1)\n",
    "    else:\n",
    "        print('Number of bases per chromosome are properly structured.')\n",
    "    \n",
    "    file=[]\n",
    "    for i in range(len(generated_sequences)):\n",
    "        file.append(str(generated_sequences[i]) + str('N')*20)      \n",
    "\n",
    "    dd = {}\n",
    "    for i in range(0,chrom_num-1):\n",
    "        dd['>sim_' + str(i + 1)] = str(file[(seq_per_chrom)*i:(seq_per_chrom)*(i+1)]).replace(\"', '\",\"\").replace(\"['\",\"\").replace(\"']\",\"\")\n",
    "    dd['>sim_' + str(chrom_num)] = str(file[(seq_per_chrom)*(chrom_num-1):]).replace(\"', '\",\"\").replace(\"['\",\"\").replace(\"']\",\"\")\n",
    "\n",
    "    df = pd.DataFrame.from_dict(dd, orient='index')\n",
    "\n",
    "    df = df.reset_index()\n",
    "    df.columns = range(df.shape[1])\n",
    "    df = df.stack()\n",
    "    df = pd.DataFrame(df).reset_index(drop=True)\n",
    "    return df.to_csv(outdir + \"/generated_sequences/\" + str(sample) + \"_simulated.fa\", header=None, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;sim_1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATCTAGACCGTAAGCGATCATGCAACTACATTGAGGGGGAATCTAG...</td>\n",
       "      <td>1531647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;sim_2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GGGGTAGGTTCCTAAATGTATTCATTTTGCAGAGGGCCTAGTTGCA...</td>\n",
       "      <td>1531647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;sim_3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GCTTAGGGGTCAAAGACGTCGCTAGAATTATATCCGGTTCCCAAGT...</td>\n",
       "      <td>1531647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>&gt;sim_4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GTGCGTCTAATCAAGCCTGAAGGTAATGGGACCCTGAATTTAAGAG...</td>\n",
       "      <td>1531647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>&gt;sim_5</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TGGGATTTTTCAGAATCGTACTATAATCGTACTTGGTCAGCTCCGT...</td>\n",
       "      <td>1531647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>&gt;sim_6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GAACAAAGGGCCACCTTTACCCAAAAAACCGAAGATGGATTATCGT...</td>\n",
       "      <td>1531647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>&gt;sim_7</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ACGCTTGGGCCCCAACAAGAAAACCCTTGGATTGTACGTACAAGCT...</td>\n",
       "      <td>1531647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>&gt;sim_8</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GTTTAGAAGGATTACTCTGGTAGCGGTGGGAACTACAAATCTTCGC...</td>\n",
       "      <td>1531647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>&gt;sim_9</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>TAAGACGGAAAAACAGACGCACACTAAATCGGATGGCGCGGGGCTA...</td>\n",
       "      <td>1531647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>&gt;sim_10</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GAGTAGGACGAAACCAGAACGCAACCAAAAAACCTGTGACACAAGT...</td>\n",
       "      <td>1540710</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  seq   counts\n",
       "0                                              >sim_1        6\n",
       "1   ATCTAGACCGTAAGCGATCATGCAACTACATTGAGGGGGAATCTAG...  1531647\n",
       "2                                              >sim_2        6\n",
       "3   GGGGTAGGTTCCTAAATGTATTCATTTTGCAGAGGGCCTAGTTGCA...  1531647\n",
       "4                                              >sim_3        6\n",
       "5   GCTTAGGGGTCAAAGACGTCGCTAGAATTATATCCGGTTCCCAAGT...  1531647\n",
       "6                                              >sim_4        6\n",
       "7   GTGCGTCTAATCAAGCCTGAAGGTAATGGGACCCTGAATTTAAGAG...  1531647\n",
       "8                                              >sim_5        6\n",
       "9   TGGGATTTTTCAGAATCGTACTATAATCGTACTTGGTCAGCTCCGT...  1531647\n",
       "10                                             >sim_6        6\n",
       "11  GAACAAAGGGCCACCTTTACCCAAAAAACCGAAGATGGATTATCGT...  1531647\n",
       "12                                             >sim_7        6\n",
       "13  ACGCTTGGGCCCCAACAAGAAAACCCTTGGATTGTACGTACAAGCT...  1531647\n",
       "14                                             >sim_8        6\n",
       "15  GTTTAGAAGGATTACTCTGGTAGCGGTGGGAACTACAAATCTTCGC...  1531647\n",
       "16                                             >sim_9        6\n",
       "17  TAAGACGGAAAAACAGACGCACACTAAATCGGATGGCGCGGGGCTA...  1531647\n",
       "18                                            >sim_10        7\n",
       "19  GAGTAGGACGAAACCAGAACGCAACCAAAAAACCTGTGACACAAGT...  1540710"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#accounting for the last chromosome where all chromosomes are the same length and the last chromosome picks up the remaining sequences\n",
    "generated_sequences= sequences_generating\n",
    "file=[]\n",
    "for i in range(len(generated_sequences)):\n",
    "    file.append(str(generated_sequences[i]) + str('N')*20)      \n",
    "\n",
    "dd = {}\n",
    "for i in range(0,chrom_num-1):\n",
    "    dd['>sim_' + str(i + 1)] = str(file[(seq_per_chrom)*i:(seq_per_chrom)*(i+1)]).replace(\"', '\",\"\").replace(\"['\",\"\").replace(\"']\",\"\")\n",
    "dd['>sim_' + str(chrom_num)] = str(file[(seq_per_chrom)*(chrom_num-1):]).replace(\"', '\",\"\").replace(\"['\",\"\").replace(\"']\",\"\")\n",
    "    \n",
    "df = pd.DataFrame.from_dict(dd, orient='index')\n",
    "\n",
    "df = df.reset_index()\n",
    "df.columns = range(df.shape[1])\n",
    "df = df.stack()\n",
    "df = pd.DataFrame(df).reset_index(drop=True)\n",
    "\n",
    "df\n",
    "# df.columns = ['seq']\n",
    "# def base_per_chromosome(row):\n",
    "#     r = pd.Series(row['seq'])\n",
    "#     x = len(row['seq'])\n",
    "#     return x\n",
    "\n",
    "# df['counts'] = df.apply(lambda row: base_per_chromosome(row), axis=1)\n",
    "# # df #1531647 on chrm 1-9\n",
    "# # #what about chrm 10\n",
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1540710"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "listf = [1,2,3,4,5,6,7,8,9,10]\n",
    "listf[4:]\n",
    "last_chrom = str(file[(seq_per_chrom)*(chrom_num-1):]).replace(\"', '\",\"\").replace(\"['\",\"\").replace(\"']\",\"\")\n",
    "len(last_chrom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Individual Sequence Length: 3021\n",
      "Number of Experimental Sequences: 5073\n",
      "Number of Chromosomes: 10\n",
      "\n",
      "Sequences per Chromosome: 507\n",
      "Sequences on Last Chromosome: 510\n",
      "Number of sequences per chromosome are properly structured.\n",
      "\n",
      "Bases per Chromosome: 1531647\n",
      "Bases on Last Chromosome: 1540710\n",
      "Total Bases in Experiment: 15325533\n",
      "Number of bases per chromosome are properly structured.\n"
     ]
    }
   ],
   "source": [
    "len(generated_sequences)\n",
    "sequence_num = 5073\n",
    "window = 1500\n",
    "chrom_num=10\n",
    "\n",
    "#sequence_num = len(file)\n",
    "seq_length = ((2*int(window))+21)\n",
    "tot_bases = int(sequence_num*seq_length)\n",
    "seq_per_chrom = math.floor(sequence_num/chrom_num)\n",
    "seq_last_chrom = seq_per_chrom + sequence_num%chrom_num\n",
    "#this is to check if there is the correct amount of sequences per chromosome\n",
    "scount = seq_last_chrom + (seq_per_chrom)*(chrom_num-1)\n",
    "\n",
    "base_per_chrom = seq_length*seq_per_chrom \n",
    "base_last_chrom = seq_length*seq_last_chrom\n",
    "#this is to check if there is the correct amount of bases per chromosome\n",
    "bscount = base_last_chrom + (base_per_chrom)*(chrom_num-1)\n",
    "\n",
    "print(\"Individual Sequence Length: \" + str(seq_length))\n",
    "print(\"Number of Experimental Sequences: \" + str(sequence_num)) \n",
    "print(\"Number of Chromosomes: \" + str(chrom_num))\n",
    "print('')\n",
    "print(\"Sequences per Chromosome: \" + str(seq_per_chrom))\n",
    "print(\"Sequences on Last Chromosome: \" + str(seq_last_chrom))\n",
    "if sequence_num-scount != 0:\n",
    "    print('Error! The total number of sequences does not equal the sum of the sequences divided into chromosomes!')\n",
    "    sys.exit(1)\n",
    "else:\n",
    "    print('Number of sequences per chromosome are properly structured.')\n",
    "print('')\n",
    "\n",
    "print(\"Bases per Chromosome: \" + str(base_per_chrom))\n",
    "print(\"Bases on Last Chromosome: \" + str(base_last_chrom))\n",
    "print(\"Total Bases in Experiment: \" + str(tot_bases))\n",
    "if tot_bases-bscount != 0:\n",
    "    print('Error! The total number of bases does not equal the sum of the bases divided into chromosomes!')\n",
    "    sys.exit(1)\n",
    "else:\n",
    "    print('Number of bases per chromosome are properly structured.')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for inserting in the python script\n",
    "def sim_bed(self):\n",
    "    #calculating the number of sequences per chromosome\n",
    "    seq_per_chrom = math.floor(self.sequence_num/self.chrom_num)\n",
    "    seq_last_chrom = seq_per_chrom + self.sequence_num%self.chrom_num\n",
    "    #creating the locations for the all chromosomes except the last\n",
    "    dbed = {}\n",
    "    for i in range(0,(seq_per_chrom)):\n",
    "        dbed['r_' + str(i)] = [int(self.window-1)+i*(2*self.window+21), int(self.window+1)+i*(2*self.window+21)]\n",
    "\n",
    "    dfbed = pd.DataFrame.from_dict(dbed, orient='index')\n",
    "    dfbed.columns = ['start', 'stop']\n",
    "    #duplicating the dataframe for all chromosomes except the last\n",
    "    dfbed = pd.concat([dfbed]*(self.chrom_num-1))\n",
    "    #adding the chromosome name to the genomic location\n",
    "    l_sim=[]\n",
    "    for i in range(int(seq_per_chrom)):\n",
    "        for c in range(int(self.chrom_num-1)):\n",
    "            l_sim.append(str('sim_' + str(c+1)))\n",
    "    l_sim = sorted(l_sim)\n",
    "    dfbed['sim'] = l_sim\n",
    "\n",
    "    #creating the locations for the last chromosome\n",
    "    dlast = {}\n",
    "    for i in range(0,(seq_last_chrom)):\n",
    "        dlast['l_' + str(i)] = [int(self.window-1)+i*(2*self.window+21), int(self.window+1)+i*(2*self.window+21)]\n",
    "    dflast = pd.DataFrame.from_dict(dlast, orient='index')\n",
    "    dflast.columns = ['start', 'stop']\n",
    "    #adding the chromosome name to the genomic location\n",
    "    l_last = []\n",
    "    for i in range(int(seq_last_chrom)):\n",
    "        l_last.append(str('sim_' + str(self.chrom_num)))\n",
    "    dflast['sim'] = l_last\n",
    "\n",
    "    #combining the bed locations\n",
    "    df = pd.concat([dfbed, dflast])\n",
    "    df = df[['sim', 'start', 'stop']]\n",
    "    #saving the new annotation\n",
    "    df.to_csv(self.outdir + \"/annotations/\" + str(self.sample) + \"_simulated_centered.bed\", header=None, index=False, sep='\\t')\n",
    "\n",
    "    #creating the windowed bed\n",
    "    df[\"start_new\"] = df.apply(lambda x: round((x[\"start\"] + x[\"stop\"])/2), axis=1)\n",
    "    df[\"stop_new\"] = df.apply(lambda x: x[\"start_new\"] + 1, axis = 1)\n",
    "\n",
    "    ##the -1500 position from \"origin\"\n",
    "    df[\"start\"] = df.apply(lambda x: x[\"start_new\"] - int(self.window), axis=1)\n",
    "\n",
    "    ##the 1500 position from the \"origin\"\n",
    "    df[\"stop\"] = df.apply(lambda x: x[\"stop_new\"] + int(self.window), axis=1)\n",
    "\n",
    "    ##saving the new annotation\n",
    "    df.to_csv(self.outdir + '/annotations/' + self.sample + '_simulated_window.bed', sep='\\t', columns=[\"sim\",\"start\",\"stop\"], header = False, index = False)\n",
    "    return print('Simulated Bed Files Genereated: ' + self.outdir + \"/annotations/\" + str(self.sample) + \"_simulated_centered.bed and \" + str(self.sample) + \"_simulated_self.window.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------Checking bed parameters------------\n",
      "Individual Sequence Length: 3021\n",
      "Number of Experimental Sequences: 5073\n",
      "Number of Chromosomes: 10\n",
      "\n",
      "Sequences per Chromosome: 507\n",
      "Sequences on Last Chromosome: 510\n",
      "Number of sequences per chromosome are properly structured.\n",
      "\n",
      "Simulated Bed Files Genereated: /Users/tajo5912/rbg/test/annotations/test_simulated_centered.bed and test_simulated_window.bed\n"
     ]
    }
   ],
   "source": [
    "##for testing in jup notebook\n",
    "def sim_bed(sequence_num, chrom_num, window, outdir, sample):\n",
    "    #calculating the number of sequences per chromosome\n",
    "    seq_per_chrom = math.floor(sequence_num/chrom_num)\n",
    "    seq_last_chrom = seq_per_chrom + sequence_num%chrom_num\n",
    "    scount = seq_last_chrom + (seq_per_chrom)*(chrom_num-1)\n",
    "    #creating the locations for the all chromosomes except the last\n",
    "    \n",
    "    print(\"----------Checking bed parameters------------\")\n",
    "    print(\"Individual Sequence Length: \" + str(seq_length))\n",
    "    print(\"Number of Experimental Sequences: \" + str(sequence_num)) \n",
    "    print(\"Number of Chromosomes: \" + str(chrom_num) +'\\n')\n",
    "    print(\"Sequences per Chromosome: \" + str(seq_per_chrom))\n",
    "    print(\"Sequences on Last Chromosome: \" + str(seq_last_chrom))\n",
    "    if sequence_num-scount != 0:\n",
    "        print('Error! The total number of sequences does not equal the sum of the sequences divided into chromosomes!')\n",
    "        sys.exit(1)\n",
    "    else:\n",
    "        print('Number of sequences per chromosome are properly structured.\\n')\n",
    "\n",
    "    dbed = {}\n",
    "    for i in range(0,(seq_per_chrom)):\n",
    "        dbed['r_' + str(i)] = [int(window-1)+i*(2*window+21), int(window+1)+i*(2*window+21)]\n",
    "\n",
    "    dfbed = pd.DataFrame.from_dict(dbed, orient='index')\n",
    "    dfbed.columns = ['start', 'stop']\n",
    "    #duplicating the dataframe for all chromosomes except the last\n",
    "    dfbed = pd.concat([dfbed]*(chrom_num-1))\n",
    "    #adding the chromosome name to the genomic location\n",
    "    l_sim=[]\n",
    "    for i in range(int(seq_per_chrom)):\n",
    "        for c in range(int(chrom_num-1)):\n",
    "            l_sim.append(str('sim_' + str(c+1)))\n",
    "    l_sim = sorted(l_sim)\n",
    "    dfbed['sim'] = l_sim\n",
    "\n",
    "    #creating the locations for the last chromosome\n",
    "    dlast = {}\n",
    "    for i in range(0,(seq_last_chrom)):\n",
    "        dlast['l_' + str(i)] = [int(window-1)+i*(2*window+21), int(window+1)+i*(2*window+21)]\n",
    "    dflast = pd.DataFrame.from_dict(dlast, orient='index')\n",
    "    dflast.columns = ['start', 'stop']\n",
    "    #adding the chromosome name to the genomic location\n",
    "    l_last = []\n",
    "    for i in range(int(seq_last_chrom)):\n",
    "        l_last.append(str('sim_' + str(chrom_num)))\n",
    "    dflast['sim'] = l_last\n",
    "\n",
    "    #combining the bed locations\n",
    "    df = pd.concat([dfbed, dflast])\n",
    "    df = df[['sim', 'start', 'stop']]\n",
    "    #saving the new annotation\n",
    "    df.to_csv(outdir + \"/annotations/\" + str(sample) + \"_simulated_centered.bed\", header=None, index=False, sep='\\t')\n",
    "\n",
    "    #creating the windowed bed\n",
    "    df[\"start_new\"] = df.apply(lambda x: round((x[\"start\"] + x[\"stop\"])/2), axis=1)\n",
    "    df[\"stop_new\"] = df.apply(lambda x: x[\"start_new\"] + 1, axis = 1)\n",
    "\n",
    "    ##the -1500 position from \"origin\"\n",
    "    df[\"start\"] = df.apply(lambda x: x[\"start_new\"] - int(window), axis=1)\n",
    "\n",
    "    ##the 1500 position from the \"origin\"\n",
    "    df[\"stop\"] = df.apply(lambda x: x[\"stop_new\"] + int(window), axis=1)\n",
    "\n",
    "    ##saving the new annotation\n",
    "    df.to_csv(outdir + '/annotations/' + sample + '_simulated_window.bed', sep='\\t', columns=[\"sim\",\"start\",\"stop\"], header = False, index = False)\n",
    "    return print('Simulated Bed Files Genereated: ' + outdir + \"/annotations/\" + str(sample) + \"_simulated_centered.bed and \" + str(sample) + \"_simulated_window.bed\")\n",
    "\n",
    "sim_bed(sequence_num=5073, chrom_num=10, window=1500, outdir='/Users/tajo5912/rbg/test', sample='test')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now to reformat the experimental \"genome\" ie the sequences based on the annotation file to generate its own genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Individual Sequence Length: 3021\n",
      "Number of Experimental Sequences: 44121\n",
      "Number of Chromosomes: 10\n",
      "\n",
      "Sequences per Chromosome: 4412\n",
      "Sequences on Last Chromosome: 4413\n",
      "Number of sequences per chromosome are properly structured.\n",
      "\n",
      "Bases per Chromosome: 13328652\n",
      "Bases on Last Chromosome: 13331673\n",
      "Total Bases in Experiment: 133289541\n",
      "Number of bases per chromosome are properly structured.\n"
     ]
    }
   ],
   "source": [
    "def reformat_expt_fasta(outdir, sample, window, chrom_num):\n",
    "    df = pd.read_csv(outdir + \"/generated_sequences/\" + sample + \"_experimental_window_sequences.fa\", header=None)\n",
    "    df = df[df.index % 2 != 0]\n",
    "    l_expt_seq = df.values.tolist()\n",
    "    flat_list = [item for sublist in l_expt_seq for item in sublist]\n",
    "    strN = str('N')*20\n",
    "    file = [\"{}{}\".format(i,strN) for i in flat_list]\n",
    "\n",
    "    sequence_num = len(file)\n",
    "    seq_length = ((2*int(window))+21)\n",
    "    tot_bases = int(sequence_num*seq_length)\n",
    "    seq_per_chrom = math.floor(sequence_num/chrom_num)\n",
    "    seq_last_chrom = seq_per_chrom + sequence_num%chrom_num\n",
    "    #this is to check if there is the correct amount of sequences per chromosome\n",
    "    scount = seq_last_chrom + (seq_per_chrom)*(chrom_num-1)\n",
    "\n",
    "    base_per_chrom = seq_length*seq_per_chrom \n",
    "    base_last_chrom = seq_length*seq_last_chrom\n",
    "    #this is to check if there is the correct amount of bases per chromosome\n",
    "    bscount = base_last_chrom + (base_per_chrom)*(chrom_num-1)\n",
    "\n",
    "    print(\"Individual Sequence Length: \" + str(seq_length))\n",
    "    print(\"Number of Experimental Sequences: \" + str(sequence_num)) \n",
    "    print(\"Number of Chromosomes: \" + str(chrom_num) +'\\n')\n",
    "    print(\"Sequences per Chromosome: \" + str(seq_per_chrom))\n",
    "    print(\"Sequences on Last Chromosome: \" + str(seq_last_chrom))\n",
    "    if sequence_num-scount != 0:\n",
    "        print('Error! The total number of sequences does not equal the sum of the sequences divided into chromosomes!')\n",
    "        sys.exit(1)\n",
    "    else:\n",
    "        print('Number of sequences per chromosome are properly structured.\\n')\n",
    "\n",
    "    print(\"Bases per Chromosome: \" + str(base_per_chrom))\n",
    "    print(\"Bases on Last Chromosome: \" + str(base_last_chrom))\n",
    "    print(\"Total Bases in Experiment: \" + str(tot_bases))\n",
    "    if tot_bases-bscount != 0:\n",
    "        print('Error! The total number of bases does not equal the sum of the bases divided into chromosomes!')\n",
    "        sys.exit(1)\n",
    "    else:\n",
    "        print('Number of bases per chromosome are properly structured.')\n",
    "\n",
    "    dd = {}\n",
    "    for i in range(0,chrom_num-1):\n",
    "        dd['>exp_' + str(i + 1)] = str(file[(seq_per_chrom)*i:(seq_per_chrom)*(i+1)]).replace(\"', '\",\"\").replace(\"['\",\"\").replace(\"']\",\"\")\n",
    "    dd['>exp_' + str(chrom_num)] = str(file[(seq_per_chrom)*(chrom_num-1):]).replace(\"', '\",\"\").replace(\"['\",\"\").replace(\"']\",\"\")\n",
    "    df = pd.DataFrame.from_dict(dd, orient='index')\n",
    "    df.columns = range(df.shape[1])\n",
    "    df=df.reset_index()\n",
    "    df = df.stack()\n",
    "    df = pd.DataFrame(df).reset_index(drop=True)\n",
    "    return df.to_csv(outdir + \"/generated_sequences/\" + str(sample) + \"_experimental.fa\", header=None, index=False, sep='\\t')\n",
    "reformat_expt_fasta('/Users/tajo5912/rbg/test', 'test', 1500,10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reformat_expt_fasta(self):\n",
    "    df = pd.read_csv(self.outdir + \"/generated_sequences/\" + self.sample + \"_experimental_self.window_sequences.fa\", header=None)\n",
    "    df = df[df.index % 2 != 0]\n",
    "    l_expt_seq = df.values.tolist()\n",
    "    flat_list = [item for sublist in l_expt_seq for item in sublist]\n",
    "    strN = str('N')*20\n",
    "    file = [\"{}{}\".format(i,strN) for i in flat_list]\n",
    "\n",
    "    sequence_num = len(file)\n",
    "    seq_length = ((2*int(self.window))+21)\n",
    "    tot_bases = int(sequence_num*seq_length)\n",
    "    seq_per_chrom = math.floor(sequence_num/self.chrom_num)\n",
    "    seq_last_chrom = seq_per_chrom + sequence_num%self.chrom_num\n",
    "    #this is to check if there is the correct amount of sequences per chromosome\n",
    "    scount = seq_last_chrom + (seq_per_chrom)*(self.chrom_num-1)\n",
    "\n",
    "    base_per_chrom = seq_length*seq_per_chrom \n",
    "    base_last_chrom = seq_length*seq_last_chrom\n",
    "    #this is to check if there is the correct amount of bases per chromosome\n",
    "    bscount = base_last_chrom + (base_per_chrom)*(self.chrom_num-1)\n",
    "\n",
    "    print(\"Individual Sequence Length: \" + str(seq_length))\n",
    "    print(\"Number of Experimental Sequences: \" + str(sequence_num)) \n",
    "    print(\"Number of Chromosomes: \" + str(self.chrom_num) +'\\n')\n",
    "    print(\"Sequences per Chromosome: \" + str(seq_per_chrom))\n",
    "    print(\"Sequences on Last Chromosome: \" + str(seq_last_chrom))\n",
    "    if sequence_num-scount != 0:\n",
    "        print('Error! The total number of sequences does not equal the sum of the sequences divided into chromosomes!')\n",
    "        sys.exit(1)\n",
    "    else:\n",
    "        print('Number of sequences per chromosome are properly structured.\\n')\n",
    "\n",
    "    print(\"Bases per Chromosome: \" + str(base_per_chrom))\n",
    "    print(\"Bases on Last Chromosome: \" + str(base_last_chrom))\n",
    "    print(\"Total Bases in Experiment: \" + str(tot_bases))\n",
    "    if tot_bases-bscount != 0:\n",
    "        print('Error! The total number of bases does not equal the sum of the bases divided into chromosomes!')\n",
    "        sys.exit(1)\n",
    "    else:\n",
    "        print('Number of bases per chromosome are properly structured.')\n",
    "\n",
    "    dd = {}\n",
    "    for i in range(0,self.chrom_num-1):\n",
    "        dd['>exp_' + str(i + 1)] = str(file[(seq_per_chrom)*i:(seq_per_chrom)*(i+1)]).replace(\"', '\",\"\").replace(\"['\",\"\").replace(\"']\",\"\")\n",
    "    dd['>exp_' + str(self.chrom_num)] = str(file[(seq_per_chrom)*(self.chrom_num-1):]).replace(\"', '\",\"\").replace(\"['\",\"\").replace(\"']\",\"\")\n",
    "    df = pd.DataFrame.from_dict(dd, orient='index')\n",
    "    df.columns = range(df.shape[1])\n",
    "    df=df.reset_index()\n",
    "    df = df.stack()\n",
    "    df = pd.DataFrame(df).reset_index(drop=True)\n",
    "    return df.to_csv(self.outdir + \"/generated_sequences/\" + str(self.sample) + \"_experimental.fa\", header=None, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------Checking bed parameters------------\n",
      "Individual Sequence Length: 3021\n",
      "Number of Experimental Sequences: 44121\n",
      "Number of Chromosomes: 10\n",
      "\n",
      "Sequences per Chromosome: 4412\n",
      "Sequences on Last Chromosome: 4413\n",
      "Number of sequences per chromosome are properly structured.\n",
      "\n",
      "Experimental Bed Files Genereated: /scratch/Users/tajo5912/rbg/levandowski2021/annotations/levandowski2021_experimental_genome_centered.bed and levandowski2021_experimental_genome_window.bed\n"
     ]
    }
   ],
   "source": [
    "##for testing in jup notebook\n",
    "def expt_bed(annotation, chrom_num, window, outdir, sample):\n",
    "    annotation_file = pd.read_csv(annotation, header=None)\n",
    "    sequence_num = len(annotation_file)\n",
    "    \n",
    "    #calculating the number of sequences per chromosome\n",
    "    seq_per_chrom = math.floor(sequence_num/chrom_num)\n",
    "    seq_last_chrom = seq_per_chrom + sequence_num%chrom_num\n",
    "    seq_length = ((2*int(window))+21)\n",
    "    tot_bases = int(sequence_num*seq_length)\n",
    "    #this is to check if there is the correct amount of sequences per chromosome\n",
    "    scount = seq_last_chrom + (seq_per_chrom)*(chrom_num-1)\n",
    "    \n",
    "    print(\"----------Checking bed parameters------------\")\n",
    "    print(\"Individual Sequence Length: \" + str(seq_length))\n",
    "    print(\"Number of Experimental Sequences: \" + str(sequence_num)) \n",
    "    print(\"Number of Chromosomes: \" + str(chrom_num) +'\\n')\n",
    "    print(\"Sequences per Chromosome: \" + str(seq_per_chrom))\n",
    "    print(\"Sequences on Last Chromosome: \" + str(seq_last_chrom))\n",
    "    if sequence_num-scount != 0:\n",
    "        print('Error! The total number of sequences does not equal the sum of the sequences divided into chromosomes!')\n",
    "        sys.exit(1)\n",
    "    else:\n",
    "        print('Number of sequences per chromosome are properly structured.\\n')\n",
    "    \n",
    "    \n",
    "    #creating the locations for the all chromosomes except the last\n",
    "    dbed = {}\n",
    "    for i in range(0,(seq_per_chrom)):\n",
    "        dbed['r_' + str(i)] = [int(window-1)+i*(2*window+21), int(window+1)+i*(2*window+21)]\n",
    "\n",
    "    dfbed = pd.DataFrame.from_dict(dbed, orient='index')\n",
    "    dfbed.columns = ['start', 'stop']\n",
    "    #duplicating the dataframe for all chromosomes except the last\n",
    "    dfbed = pd.concat([dfbed]*(chrom_num-1))\n",
    "    #adding the chromosome name to the genomic location\n",
    "    l_expt=[]\n",
    "    for i in range(int(seq_per_chrom)):\n",
    "        for c in range(int(chrom_num-1)):\n",
    "            l_expt.append(str('exp_' + str(c+1)))\n",
    "    l_expt = sorted(l_expt)\n",
    "    dfbed['expt'] = l_expt\n",
    "\n",
    "    #creating the locations for the last chromosome\n",
    "    dlast = {}\n",
    "    for i in range(0,(seq_last_chrom)):\n",
    "        dlast['l_' + str(i)] = [int(window-1)+i*(2*window+21), int(window+1)+i*(2*window+21)]\n",
    "    dflast = pd.DataFrame.from_dict(dlast, orient='index')\n",
    "    dflast.columns = ['start', 'stop']\n",
    "    #adding the chromosome name to the genomic location\n",
    "    l_last = []\n",
    "    for i in range(int(seq_last_chrom)):\n",
    "        l_last.append(str('exp_' + str(chrom_num)))\n",
    "    dflast['expt'] = l_last\n",
    "\n",
    "    #combining the bed locations\n",
    "    df = pd.concat([dfbed, dflast])\n",
    "    df = df[['expt', 'start', 'stop']]\n",
    "    #saving the new annotation\n",
    "    df.to_csv(outdir + \"/annotations/\" + str(sample) + \"_experimental_genome_centered.bed\", header=None, index=False, sep='\\t')\n",
    "\n",
    "    #creating the windowed bed\n",
    "    df[\"start_new\"] = df.apply(lambda x: round((x[\"start\"] + x[\"stop\"])/2), axis=1)\n",
    "    df[\"stop_new\"] = df.apply(lambda x: x[\"start_new\"] + 1, axis = 1)\n",
    "\n",
    "    ##the -1500 position from \"origin\"\n",
    "    df[\"start\"] = df.apply(lambda x: x[\"start_new\"] - int(window), axis=1)\n",
    "\n",
    "    ##the 1500 position from the \"origin\"\n",
    "    df[\"stop\"] = df.apply(lambda x: x[\"stop_new\"] + int(window), axis=1)\n",
    "    \n",
    "    ##saving the new annotation\n",
    "    df.to_csv(outdir + '/annotations/' + sample + '_experimental_genome_window.bed', sep='\\t', columns=[\"expt\",\"start\",\"stop\"], header = False, index = False)    \n",
    "    return print('Experimental Bed Files Genereated: ' + outdir + \"/annotations/\" + str(sample) + \"_experimental_genome_centered.bed and \" + str(sample) + \"_experimental_genome_window.bed\")\n",
    "\n",
    "expt_bed(annotation='/scratch/Users/tajo5912/rbg/levandow_mumerge_bidir.bed', chrom_num=10, window=1500, outdir='/scratch/Users/tajo5912/rbg/levandowski2021', sample='levandowski2021')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44121"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outdir = '/Users/tajo5912/rbg/test'\n",
    "annotation = '/scratch/Users/tajo5912/rbg/levandow_mumerge_bidir.bed'\n",
    "sample='test'\n",
    "df = pd.read_csv(annotation, header=None)\n",
    "len(df) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##for main\n",
    "def expt_bed(self):\n",
    "    annotation_file = pd.read_csv(self.annotation, header=None)\n",
    "    sequence_num = len(annotation_file)\n",
    "    \n",
    "    #calculating the number of sequences per chromosome\n",
    "    seq_per_chrom = math.floor(sequence_num/self.chrom_num)\n",
    "    seq_last_chrom = seq_per_chrom + sequence_num%self.chrom_num\n",
    "    scount = seq_last_chrom + (seq_per_chrom)*(self.chrom_num-1)\n",
    "    \n",
    "    print(\"----------Checking bed parameters------------\")\n",
    "    print(\"Individual Sequence Length: \" + str(seq_length))\n",
    "    print(\"Number of Experimental Sequences: \" + str(sequence_num)) \n",
    "    print(\"Number of Chromosomes: \" + str(self.chrom_num) +'\\n')\n",
    "    print(\"Sequences per Chromosome: \" + str(seq_per_chrom))\n",
    "    print(\"Sequences on Last Chromosome: \" + str(seq_last_chrom))\n",
    "    if sequence_num-scount != 0:\n",
    "        print('Error! The total number of sequences does not equal the sum of the sequences divided into chromosomes!')\n",
    "        sys.exit(1)\n",
    "    else:\n",
    "        print('Number of sequences per chromosome are properly structured.\\n')\n",
    "    \n",
    "    \n",
    "    #creating the locations for the all chromosomes except the last\n",
    "    dbed = {}\n",
    "    for i in range(0,(seq_per_chrom)):\n",
    "        dbed['r_' + str(i)] = [int(self.window-1)+i*(2*self.window+21), int(self.window+1)+i*(2*self.window+21)]\n",
    "\n",
    "    dfbed = pd.DataFrame.from_dict(dbed, orient='index')\n",
    "    dfbed.columns = ['start', 'stop']\n",
    "    #duplicating the dataframe for all chromosomes except the last\n",
    "    dfbed = pd.concat([dfbed]*(self.chrom_num-1))\n",
    "    #adding the chromosome name to the genomic location\n",
    "    l_expt=[]\n",
    "    for i in range(int(seq_per_chrom)):\n",
    "        for c in range(int(self.chrom_num-1)):\n",
    "            l_expt.append(str('expt_' + str(c+1)))\n",
    "    l_expt = sorted(l_expt)\n",
    "    dfbed['expt'] = l_expt\n",
    "\n",
    "    #creating the locations for the last chromosome\n",
    "    dlast = {}\n",
    "    for i in range(0,(seq_last_chrom)):\n",
    "        dlast['l_' + str(i)] = [int(self.window-1)+i*(2*self.window+21), int(self.window+1)+i*(2*self.window+21)]\n",
    "    dflast = pd.DataFrame.from_dict(dlast, orient='index')\n",
    "    dflast.columns = ['start', 'stop']\n",
    "    #adding the chromosome name to the genomic location\n",
    "    l_last = []\n",
    "    for i in range(int(seq_last_chrom)):\n",
    "        l_last.append(str('expt_' + str(self.chrom_num)))\n",
    "    dflast['expt'] = l_last\n",
    "\n",
    "    #combining the bed locations\n",
    "    df = pd.concat([dfbed, dflast])\n",
    "    df = df[['expt', 'start', 'stop']]\n",
    "    #saving the new annotation\n",
    "    df.to_csv(self.outdir + \"/annotations/\" + str(self.sample) + \"_experimental_genome_centered.bed\", header=None, index=False, sep='\\t')\n",
    "\n",
    "    #creating the self.windowed bed\n",
    "    df[\"start_new\"] = df.apply(lambda x: round((x[\"start\"] + x[\"stop\"])/2), axis=1)\n",
    "    df[\"stop_new\"] = df.apply(lambda x: x[\"start_new\"] + 1, axis = 1)\n",
    "\n",
    "    ##the -1500 position from \"origin\"\n",
    "    df[\"start\"] = df.apply(lambda x: x[\"start_new\"] - int(self.window), axis=1)\n",
    "\n",
    "    ##the 1500 position from the \"origin\"\n",
    "    df[\"stop\"] = df.apply(lambda x: x[\"stop_new\"] + int(self.window), axis=1)\n",
    "    \n",
    "    ##saving the new annotation\n",
    "    df.to_csv(self.outdir + '/annotations/' + self.sample + '_experimental_genome_self.window.bed', sep='\\t', columns=[\"expt\",\"start\",\"stop\"], header = False, index = False)    \n",
    "    return print('Experimental Bed Files Genereated: ' + self.outdir + \"/annotations/\" + str(self.sample) + \"_experimental_genome_centered.bed and \" + str(self.sample) + \"_experimental_genome_self.window.bed\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('/scratch/Users/tajo5912/rbg/levandowski2021/annotations/levandowski2021_simulated_centered.bed', header= None, sep='\\t')\n",
    "\n",
    "\n",
    "df=df.iloc[0:5001]\n",
    "df.to_csv('/Users/tajo5912/rbg/md_testing/5000_sim.bed', header=None, index=False, sep='\\t')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
