{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from os import path\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "red=False\n",
    "blue=False\n",
    "if red == True:\n",
    "    print('red is true')\n",
    "elif blue == True:\n",
    "    print('blue is true')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "red=True\n",
    "blue=False\n",
    "if (red==True) and (blue==True):\n",
    "    raise ValueError(\"cant be both red and blue\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1.950000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1.589704e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>4.680446e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>9.700000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.997000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>3.203200e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.667610e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.489564e+08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             counts\n",
       "count  1.950000e+02\n",
       "mean   1.589704e+07\n",
       "std    4.680446e+07\n",
       "min    9.700000e+02\n",
       "25%    1.997000e+03\n",
       "50%    3.203200e+04\n",
       "75%    1.667610e+05\n",
       "max    2.489564e+08"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chs = pd.read_csv('/scratch/Shares/dowell/genomes/hg38/hg38.chrom.sizes', sep='\\t', header=None)\n",
    "chs.columns = ['chr', 'counts']\n",
    "chs['counts'].mean()\n",
    "chs.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fimotobed(outdir):\n",
    "    def identifier(row):\n",
    "        return str(row['sequence_name']) + ':' + str(row['start']) + '-' + str(row['stop']) + '_' + str(row['strand'] + ';' + str(row['motif_id']))\n",
    "\n",
    "    def percent_gc(row):\n",
    "        r = pd.Series(row['matched_sequence'])\n",
    "        z = r.str.count('G') + r.str.count('C')\n",
    "        x = z/len(row['matched_sequence'])\n",
    "        return x\n",
    "\n",
    "    motif_dirs = glob.glob(outdir + '/temp/sim_fimo_out/*')\n",
    "    for motif_dir in motif_dirs:\n",
    "        if path.isdir(motif_dir):\n",
    "            motif_name = motif_dir.split('/')[-1]\n",
    "            print(motif_name)\n",
    "            df = pd.read_csv('%s/fimo.tsv' % (motif_dir), sep ='\\t')\n",
    "            print(\"Post-processing FIMO output for %s\" % motif_name)\n",
    "            df.drop(df.tail(3).index,inplace=True)\n",
    "            if df.empty == True:\n",
    "                print('Skipping ' + motif_name + ' -no motif hits.')\n",
    "            else:\n",
    "                df = df.sort_values('sequence_name').reset_index()\n",
    "                df['identifier'] = df.apply(lambda row: identifier(row), axis=1)\n",
    "                df['percent_gc'] = df.apply(lambda row: percent_gc(row), axis=1)\n",
    "                df['mean_percent_gc'] = df['percent_gc'].mean()\n",
    "                df = df[['sequence_name', 'start', 'stop', 'identifier', 'score', 'strand', \n",
    "                         'matched_sequence', 'percent_gc', 'mean_percent_gc']]\n",
    "                df.to_csv(outdir + '/motifs/simulated/' + motif_name + '.sorted.bed', sep='\\t', header=None, index=False)\n",
    "\n",
    "def gc_content(outdir):\n",
    "    directory = os.fsencode(outdir + '/motifs/simulated')\n",
    "    output = pd.DataFrame({'identifier' : [], 'mean_percent_gc' : []})\n",
    "    for file in os.listdir(directory):\n",
    "        if file.endswith(b'.bed'):\n",
    "            file = file.decode('utf-8')\n",
    "            df = pd.read_csv(outdir + '/motifs/simulated/' + file, sep=\"\\t\", header = None, \n",
    "                             names = ['sequence_name', 'start', 'stop', 'identifier', 'score', 'strand', 'matched_sequence', 'percent_gc', 'mean_percent_gc'])\n",
    "            df = df[['identifier', 'mean_percent_gc']].loc[0]\n",
    "            output=output.append(df)\n",
    "            output.to_csv(outdir + '/annotations/simulated_mean_gc_percent.txt', sep=\"\\t\", header=None, index=False)\n",
    "        else:\n",
    "            continue\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fimotobed(outdir = '/scratch/Users/tajo5912/rbg/test')\n",
    "gc_content(outdir = '/scratch/Users/tajo5912/rbg/test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Testing functions for fimo formating\n",
    "# fimoout = '/scratch/Users/tajo5912/rbg/test/temp/fimo'\n",
    "# bedout = '/scratch/Users/tajo5912/rbg/test/temp/bed'\n",
    "\n",
    "# def identifier(row):\n",
    "#     return str(row['sequence_name']) + ':' + str(row['start']) + '-' + str(row['stop']) + '_' + str(row['strand'] + ';' + str(row['motif_id']))\n",
    "\n",
    "# def percent_gc(row):\n",
    "#     r = pd.Series(row['matched_sequence'])\n",
    "#     z = r.str.count('G') + r.str.count('C')\n",
    "#     x = z/len(row['matched_sequence'])\n",
    "#     return x\n",
    "\n",
    "# motif_dirs = glob.glob(fimoout + '/*')\n",
    "# for motif_dir in motif_dirs:\n",
    "#     if path.isdir(motif_dir):\n",
    "#         motif_name = motif_dir.split('/')[-1]\n",
    "#         print(motif_name)\n",
    "#         df = pd.read_csv('%s/fimo.tsv' % (motif_dir), sep ='\\t')\n",
    "#         print(\"Post-processing FIMO output for %s\" % motif_name)\n",
    "#         df.drop(df.tail(3).index,inplace=True)\n",
    "#         if df.empty == True:\n",
    "#             print('Skipping ' + motif_name + ' -no motif hits.')\n",
    "#         else:\n",
    "#             df = df.sort_values('sequence_name').reset_index()\n",
    "#             df['identifier'] = df.apply(lambda row: identifier(row), axis=1)\n",
    "#             df['percent_gc'] = df.apply(lambda row: percent_gc(row), axis=1)\n",
    "#             df['mean_percent_gc'] = df['percent_gc'].mean()\n",
    "#             df = df[['sequence_name', 'start', 'stop', 'identifier', 'score', 'strand', \n",
    "#                      'matched_sequence', 'percent_gc', 'mean_percent_gc']]\n",
    "#             df.to_csv(bedout + '/' + motif_name + '.sorted.bed', sep='\\t', header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skip\n"
     ]
    }
   ],
   "source": [
    "gata = '/scratch/Users/tajo5912/rbg/test/temp/fimo/GATA3_M05752_1/fimo.tsv'\n",
    "gata = pd.read_csv(gata, sep='\\t')\n",
    "gata.drop(gata.tail(3).index,inplace=True)\n",
    "if gata.empty == True:\n",
    "    print('skip')\n",
    "else:\n",
    "    df['identifier'] = df.apply(lambda row: identifier(row), axis=1)\n",
    "    df['percent_gc'] = df.apply(lambda row: percent_gc(row), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "outdir='/scratch/Users/tajo5912/rbg/test'\n",
    "directory = os.fsencode(outdir + '/motifs/simulated')\n",
    "output = pd.DataFrame({'identifier' : [], 'mean_percent_gc' : []})\n",
    "for file in os.listdir(directory):\n",
    "    if file.endswith(b'.bed'):\n",
    "        file = file.decode('utf-8')\n",
    "        df = pd.read_csv(outdir + '/motifs/simulated/' + file, sep=\"\\t\", header = None, \n",
    "                         names = ['sequence_name', 'start', 'stop', 'identifier', 'score', 'strand', 'matched_sequence', 'percent_gc', 'mean_percent_gc'])\n",
    "        df = df[['identifier', 'mean_percent_gc']].loc[0]\n",
    "        output=output.append(df)\n",
    "        output.to_csv(outdir + '/annotations/simulated_mean_gc_percent.txt', sep=\"\\t\", header=None, index=False)\n",
    "    else:\n",
    "        continue\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hg38\n",
      "Successfully created chromosome size file /scratch/Users/tajo5912/rbg/test/generated_sequences/hg38.chrom.sizes\n"
     ]
    }
   ],
   "source": [
    "def chrm_sizes_whole_genome(genome, outdir):\n",
    "    g = genome.split('/')[-1].split('.')[-2]\n",
    "    print(g)\n",
    "    try:\n",
    "        os.system(\"cut -f1,2 \" + genome + \".fai > \" + outdir + \"/generated_sequences/\" + g + \".chrom.sizes\")\n",
    "    except OSError:\n",
    "        print (\"Creation of chromosome size file %s failed\" % outdir + \"/generated_sequences/\" + g + \".chrom.sizes\")\n",
    "    else:\n",
    "        print (\"Successfully created chromosome size file %s\" % outdir + \"/generated_sequences/\" + g + \".chrom.sizes\")\n",
    "            \n",
    "chrm_sizes_whole_genome(genome='/scratch/Shares/dowell/genomes/hg38/hg38.fa', \n",
    "                        outdir='/scratch/Users/tajo5912/rbg/test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "os.stat_result(st_mode=33188, st_ino=9088365749575586314, st_dev=41, st_nlink=1, st_uid=573627, st_gid=1100000, st_size=103100, st_atime=1612546485, st_mtime=1612546485, st_ctime=1612546485)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.stat('/scratch/Users/tajo5912/rbg/test/annotations/test_simulated_window.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Adapted from DAStk by Ignacio\n",
    "\n",
    "\n",
    "import csv\n",
    "import datetime\n",
    "import glob\n",
    "import multiprocessing\n",
    "import numpy as np\n",
    "import os.path\n",
    "import pandas as pd\n",
    "import sys\n",
    "import matplotlib as mpl\n",
    "import pybedtools\n",
    "# to prevent DISPLAY weirdness when running in the cluster:\n",
    "mpl.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "plt.ioff()\n",
    "from functools import partial\n",
    "from operator import itemgetter\n",
    "#from functools import reduce\n",
    "\n",
    "\n",
    "# around the given mean\n",
    "def is_in_window(motif_interval, bed_median, window_size):\n",
    "    start = int(motif_interval.start)\n",
    "    end = int(motif_interval.end)\n",
    "    if (end >= (bed_median - window_size) and end <= (bed_median + window_size)) \\\n",
    "        or (start >= (bed_median - window_size) and start <= (bed_median + window_size)):\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "    \n",
    "\n",
    "\n",
    "def find_motifs_in_chrom(current_chrom, files):\n",
    "    tf_motif_filename, windowbed_name, window = files\n",
    "    H = window          # in bps, the MD-score parameter (large window)\n",
    "    h = window * 0.1           # in bps, the MD-score parameter (small window)\n",
    "\n",
    "    wdf = pd.read_csv(windowbed_name, header=None, comment='#', sep=\"\\t\", usecols=[0, 1, 2], \\\n",
    "                          names=['chrom', 'start', 'end'], na_filter=False, dtype={'chrom':'str', 'start':'int', 'end':'int'})\n",
    "    w_iter = wdf[(wdf.chrom == current_chrom)].itertuples()\n",
    "    motif_df = pd.read_csv(tf_motif_filename, header=None, comment='#', sep=\"\\t\", usecols=[0, 1, 2], \\\n",
    "                           names=['chrom', 'start', 'end'], na_filter=False, dtype={'chrom':'str', 'start':'int', 'end':'int'})\n",
    "    if len(motif_df) == 0:\n",
    "        return None\n",
    "    motif_iter = motif_df[(motif_df.chrom == current_chrom)].itertuples()\n",
    "    last_motif = None\n",
    "    g_h = 0\n",
    "    g_H = 0\n",
    "    total_motif_sites = 0\n",
    "    tf_distances = []\n",
    "    try:\n",
    "        motif_region = next(motif_iter)   # get the first motif in the list\n",
    "    except StopIteration:\n",
    "        print('No motifs for chromosome ' + current_chrom + ' on file ' + tf_motif_filename)\n",
    "        return None\n",
    "\n",
    "    peak_count_overlapping_motif = 0\n",
    "    for windowbed_peak in atac_iter:\n",
    "        motifs_within_region = True\n",
    "\n",
    "        bed_median = bed_peak.start + (bed_peak.end - bed_peak.start)/2\n",
    "        # check the last motif, too, in case any of them falls within the region of\n",
    "        # interest of two sequential ATAC-Seq peaks\n",
    "        if last_motif:\n",
    "            if is_in_window(last_motif, bed_median, h):\n",
    "                g_h = g_h + 1\n",
    "            if is_in_window(last_motif, bed_median, H):\n",
    "                g_H = g_H + 1\n",
    "                tf_median = last_motif.start + (last_motif.end - last_motif.start)/2\n",
    "                tf_distances.append(bed_median - tf_median)\n",
    "                try:\n",
    "                    motif_region = next(motif_iter)   # get the next motif in the list\n",
    "                    total_motif_sites += 1\n",
    "                except StopIteration:\n",
    "                    pass\n",
    "                last_motif = motif_region\n",
    "                if motif_region.start > (bed_median + H):\n",
    "                    motifs_within_region = False\n",
    "\n",
    "        # Move to the next putative motif sites until we get one past our evaluation window\n",
    "        while motifs_within_region:\n",
    "            total_motif_sites += 1\n",
    "            # account for those within the smaller window (h)\n",
    "            if is_in_window(motif_region, bed_median, h):\n",
    "                g_h = g_h + 1\n",
    "            # account for those within the larger window (H)\n",
    "            if is_in_window(motif_region, bed_median, H):\n",
    "                g_H = g_H + 1\n",
    "                tf_median = motif_region.start + (motif_region.end - motif_region.start)/2\n",
    "                tf_distances.append(bed_median - tf_median)\n",
    "\n",
    "            # if we still haven't shifted past this peak...\n",
    "            if motif_region.start <= (bed_median + H):\n",
    "                try:\n",
    "                    motif_region = next(motif_iter)   # get the next motif in the list\n",
    "                    total_motif_sites += 1\n",
    "                except StopIteration:\n",
    "                    # No more TF motifs for this chromosome\n",
    "                    break\n",
    "                last_motif = motif_region\n",
    "            else:\n",
    "                motifs_within_region = False\n",
    "\n",
    "    # Count any remaining TF motif sites after the last peak\n",
    "    while(len(motif_region) > 0):\n",
    "        try:\n",
    "            motif_region = next(motif_iter)   # this gets the next motif in the list\n",
    "        except StopIteration:\n",
    "            break\n",
    "\n",
    "    return [tf_distances, g_h, g_H, total_motif_sites]\n",
    "\n",
    "\n",
    "#md_score, small_window, large_window, motif_site_count, heat] = \n",
    "#et_md_score(filename, int(threads), windowbed_name, CHROMOSOMES, window)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MD functions .py\n",
    "#filename, int(threads), windowbed_name, CHROMOSOMES, window\n",
    "# around the given mean\n",
    "def is_in_window(motif_interval, bed_median, window_size):\n",
    "    start = int(motif_interval.start)\n",
    "    end = int(motif_interval.end)\n",
    "    if (end >= (bed_median - window_size) and end <= (bed_median + window_size)) \\\n",
    "        or (start >= (bed_median - window_size) and start <= (bed_median + window_size)):\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "    \n",
    "\n",
    "\n",
    "def find_motifs_in_chrom(current_chrom, files):\n",
    "    tf_motif_filename, windowbed_name, window = files\n",
    "    H = window          # in bps, the MD-score parameter (large window)\n",
    "    h = window * 0.1           # in bps, the MD-score parameter (small window)\n",
    "\n",
    "    wdf = pd.read_csv(windowbed_name, header=None, comment='#', sep=\"\\t\", usecols=[0, 1, 2], \\\n",
    "                          names=['chrom', 'start', 'end'], na_filter=False, dtype={'chrom':'str', 'start':'int', 'end':'int'})\n",
    "    w_iter = wdf[(wdf.chrom == current_chrom)].itertuples()\n",
    "    motif_df = pd.read_csv(tf_motif_filename, header=None, comment='#', sep=\"\\t\", usecols=[0, 1, 2], \\\n",
    "                           names=['chrom', 'start', 'end'], na_filter=False, dtype={'chrom':'str', 'start':'int', 'end':'int'})\n",
    "    if len(motif_df) == 0:\n",
    "        return None\n",
    "    motif_iter = motif_df[(motif_df.chrom == current_chrom)].itertuples()\n",
    "    last_motif = None\n",
    "    g_h = 0\n",
    "    g_H = 0\n",
    "    total_motif_sites = 0\n",
    "    tf_distances = []\n",
    "    try:\n",
    "        motif_region = next(motif_iter)   # get the first motif in the list\n",
    "    except StopIteration:\n",
    "        print('No motifs for chromosome ' + current_chrom + ' on file ' + tf_motif_filename)\n",
    "        return None\n",
    "\n",
    "    peak_count_overlapping_motif = 0\n",
    "    for windowbed_peak in atac_iter:\n",
    "        motifs_within_region = True\n",
    "\n",
    "        bed_median = bed_peak.start + (bed_peak.end - bed_peak.start)/2\n",
    "        # check the last motif, too, in case any of them falls within the region of\n",
    "        # interest of two sequential ATAC-Seq peaks\n",
    "        if last_motif:\n",
    "            if is_in_window(last_motif, bed_median, h):\n",
    "                g_h = g_h + 1\n",
    "            if is_in_window(last_motif, bed_median, H):\n",
    "                g_H = g_H + 1\n",
    "                tf_median = last_motif.start + (last_motif.end - last_motif.start)/2\n",
    "                tf_distances.append(bed_median - tf_median)\n",
    "                try:\n",
    "                    motif_region = next(motif_iter)   # get the next motif in the list\n",
    "                    total_motif_sites += 1\n",
    "                except StopIteration:\n",
    "                    pass\n",
    "                last_motif = motif_region\n",
    "                if motif_region.start > (bed_median + H):\n",
    "                    motifs_within_region = False\n",
    "\n",
    "        # Move to the next putative motif sites until we get one past our evaluation window\n",
    "        while motifs_within_region:\n",
    "            total_motif_sites += 1\n",
    "            # account for those within the smaller window (h)\n",
    "            if is_in_window(motif_region, bed_median, h):\n",
    "                g_h = g_h + 1\n",
    "            # account for those within the larger window (H)\n",
    "            if is_in_window(motif_region, bed_median, H):\n",
    "                g_H = g_H + 1\n",
    "                tf_median = motif_region.start + (motif_region.end - motif_region.start)/2\n",
    "                tf_distances.append(bed_median - tf_median)\n",
    "\n",
    "            # if we still haven't shifted past this peak...\n",
    "            if motif_region.start <= (bed_median + H):\n",
    "                try:\n",
    "                    motif_region = next(motif_iter)   # get the next motif in the list\n",
    "                    total_motif_sites += 1\n",
    "                except StopIteration:\n",
    "                    # No more TF motifs for this chromosome\n",
    "                    break\n",
    "                last_motif = motif_region\n",
    "            else:\n",
    "                motifs_within_region = False\n",
    "\n",
    "    # Count any remaining TF motif sites after the last peak\n",
    "    while(len(motif_region) > 0):\n",
    "        try:\n",
    "            motif_region = next(motif_iter)   # this gets the next motif in the list\n",
    "        except StopIteration:\n",
    "            break\n",
    "\n",
    "    return [tf_distances, g_h, g_H, total_motif_sites]\n",
    "\n",
    "def get_md_score(tf_motif_filename, threads, windowbed_name, CHROMOSOMES, window):\n",
    "\n",
    "    HISTOGRAM_BINS = 150\n",
    "    pool = multiprocessing.Pool(threads)\n",
    "    results = pool.map(partial(find_motifs_in_chrom, files=[tf_motif_filename, windowbed_name, window]), CHROMOSOMES)\n",
    "    pool.close()\n",
    "    pool.join()\n",
    "\n",
    "    results_matching_motif = [x for x in results if x is not None]\n",
    "    if len(results_matching_motif) > 0:\n",
    "        sums = np.sum(results_matching_motif, axis=0)\n",
    "        overall_g_h = sums[1]\n",
    "        overall_g_H = sums[2]\n",
    "        overall_motif_sites = sums[3]\n",
    "\n",
    "        # Calculate the heatmap for this motif's barcode\n",
    "        tf_distances = sums[0]\n",
    "        heatmap, xedges = np.histogram(tf_distances, bins=HISTOGRAM_BINS)\n",
    "        str_heatmap = np.char.mod('%d', heatmap)\n",
    "        if overall_g_H >= 0:\n",
    "            return [float(overall_g_h + 1)/(overall_g_H + 1), (overall_g_h + 1), (overall_g_H + 1), (overall_motif_sites + 1), ';'.join(str_heatmap)]\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ### Run this as main, drop MD functions and only use MDScore class for pooling?\n",
    "# import os\n",
    "# import sys\n",
    "# from os import path\n",
    "# import pandas as pd\n",
    "# import csv\n",
    "\n",
    "# class MDScore:\n",
    "#     '''Setting up for MD-Score calculations'''\n",
    "#     def __init__(self, window, threads, outdir, sample):\n",
    "#         self.window = window\n",
    "#         self.threads = threads\n",
    "#         self.outdir = outdir\n",
    "#         self.sample = sample   \n",
    "\n",
    "#     def mdcalc(self):\n",
    "#         windowbed_name = self.outdir + '/annotations/' + self.sample + '_simulated_window.bed' \n",
    "#         input_file_statinfo = os.stat(windowbed_name)    \n",
    "#         if input_file_statinfo.st_size !=0:\n",
    "#             windowbed = open(windowbed_name)\n",
    "#             windowbed_csv_reader = csv.reader(windowbed, delimiter='\\t')\n",
    "#             windowbed_line = next(windowbed_csv_reader)\n",
    "#             windowbed_peak_count = 0        \n",
    "#         else:\n",
    "#             raise ValueError(\"The windowed annotation file (sample_simulated_window.bed) is empty!\")\n",
    "        \n",
    "#         chromosomes = self.outdir + '/generated_sequences/' + self.sample + '_simulated.chrom.sizes'\n",
    "#         if chromosomes:\n",
    "#             chr_df = pd.read_csv(chromosomes, header=None, comment='#', sep=\"\\t\", usecols=[0,1], \\\n",
    "#                                       names=['chrom', 'size'], na_filter=False, dtype={'chrom':'str', 'size':'int'})\n",
    "#             chr_list = list(chr_df['chrom'])\n",
    "#             CHROMOSOMES = [word for word in chr_list if len(word) <= 6]\n",
    "#         else:\n",
    "#             raise ValueError(\"Chromosome size file missing from generated_sequences(sample_simulated.chrom.sizes)!\") \n",
    "            \n",
    "#         windowbed.seek(0)\n",
    "#         windowbed_csv_reader = csv.reader(windowbed, delimiter='\\t')\n",
    "#         windowbed_line = next(windowbed_csv_reader)\n",
    "#         while(windowbed_line[0][0] == '#'):\n",
    "#             windowbed_line = next(windowbed_csv_reader)    \n",
    "        \n",
    "#         motif_stats = []\n",
    "#         sorted_motif_stats = []\n",
    "#         tf_motif_path = self.outdir + '/motifs/simulated/*'\n",
    "#         motif_filenames = glob.glob(tf_motif_path)\n",
    "#         motif_count = len(motif_filenames)\n",
    "#         print(\"Processing %s motif files in %s\" % motif_count, tf_motif_path)\n",
    "#         for filename in motif_filenames:\n",
    "#             filename_no_path = filename.split('/')[-1]\n",
    "#             if os.path.getsize(filename) > 0 and os.path.basename(filename).endswith(tuple(['.bed'])):\n",
    "#                 [md_score, small_window, large_window, motif_site_count, heat] = get_md_score(filename, int(self.threads), windowbed_name, CHROMOSOMES, self.window)\n",
    "#                 print('The MD-score for the provided BED regions vs %s is %.6f' % (filename_no_path, md_score))\n",
    "#                 motif_stats.append({ 'motif_file': filename_no_path,'md_score': md_score, 'small_window': small_window, 'large_window': large_window, 'motif_site_count': motif_site_count, 'heat': heat })\n",
    "#         # sort the stats dictionary by MD-score, descending order\n",
    "#         sorted_motif_stats = sorted(motif_stats, key=itemgetter('md_score'), reverse=True)\n",
    "\n",
    "#         md_score_fp = open(\"%s/final_output/%s_simulated_md_scores.txt\" % (self.outdir, self.sample), 'w')\n",
    "#         for stat in sorted_motif_stats:\n",
    "#             md_score_fp.write(\"%s,%s,%s,%s,%s,%s\\n\" % (stat['motif_file'], stat['md_score'], stat['small_window'], \\\n",
    "#                                stat['large_window'], stat['motif_site_count'], stat['heat']))\n",
    "#         md_score_fp.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Individual Sequence Length: 3021\n",
      "Number of Experimental Sequences: 5073\n",
      "Number of Chromosomes: 10\n",
      "\n",
      "Sequences per Chromosome: 507\n",
      "Sequences on Last Chromosome: 510\n",
      "\n",
      "Bases per Chromosome: 1531647\n",
      "Bases on Last Chromosome: 1540710\n",
      "Total Bases in Experiment: 15325533\n"
     ]
    },
    {
     "ename": "SyntaxError",
     "evalue": "'continue' not properly in loop (<ipython-input-87-12598edc156e>, line 34)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-87-12598edc156e>\"\u001b[0;36m, line \u001b[0;32m34\u001b[0m\n\u001b[0;31m    continue\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m 'continue' not properly in loop\n"
     ]
    }
   ],
   "source": [
    "sequence_num = 5073\n",
    "window = 1500\n",
    "chrom_num=10\n",
    "\n",
    "#sequence_num = len(file)\n",
    "seq_length = ((2*int(window))+21)\n",
    "tot_bases = int(sequence_num*seq_length)\n",
    "\n",
    "seq_per_chrom = math.floor(sequence_num/chrom_num)\n",
    "seq_last_chrom = seq_per_chrom + sequence_num%chrom_num\n",
    "base_per_chrom = seq_length*seq_per_chrom \n",
    "base_last_chrom = seq_length*seq_last_chrom\n",
    "bscount = base_last_chrom + (base_per_chrom)*(chrom_num-1)\n",
    "\n",
    "\n",
    "bs_tot_seq = seq_last_chrom + seq_per_chrom*(chrom_num-1)\n",
    "\n",
    "print(\"Individual Sequence Length: \" + str(seq_length))\n",
    "print(\"Number of Experimental Sequences: \" + str(sequence_num)) \n",
    "print(\"Number of Chromosomes: \" + str(chrom_num))\n",
    "print('')\n",
    "\n",
    "print(\"Sequences per Chromosome: \" + str(seq_per_chrom))\n",
    "print(\"Sequences on Last Chromosome: \" + str(seq_last_chrom))\n",
    "print('')\n",
    "\n",
    "print(\"Bases per Chromosome: \" + str(base_per_chrom))\n",
    "print(\"Bases on Last Chromosome: \" + str(base_last_chrom))\n",
    "print(\"Total Bases in Experiment: \" + str(tot_bases))\n",
    "if tot_bases-bscount != 0:\n",
    "    print('Error! The total number of bases does not equal the sum of the bases divided into chromosomes!')\n",
    "    sys.exit(1)\n",
    "else:\n",
    "    continue\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Flattening classes for main\n",
    "### Run this as the func in main that calls in md_functions\n",
    "#for simulated only !\n",
    "import os\n",
    "import sys\n",
    "from os import path\n",
    "import pandas as pd\n",
    "import csv\n",
    "\n",
    "def mdcalc(window, threads, outdir, sample):\n",
    "    windowbed_name = outdir + '/annotations/' + sample + '_simulated_window.bed' \n",
    "    input_file_statinfo = os.stat(windowbed_name)    \n",
    "    if input_file_statinfo.st_size !=0:\n",
    "        windowbed = open(windowbed_name)\n",
    "        windowbed_csv_reader = csv.reader(windowbed, delimiter='\\t')\n",
    "        windowbed_line = next(windowbed_csv_reader)\n",
    "        windowbed_peak_count = 0        \n",
    "    else:\n",
    "        raise ValueError(\"The windowed annotation file (sample_simulated_window.bed) is empty!\")\n",
    "\n",
    "    chromosomes = outdir + '/generated_sequences/' + sample + '_simulated.chrom.sizes'\n",
    "    if chromosomes:\n",
    "        chr_df = pd.read_csv(chromosomes, header=None, comment='#', sep=\"\\t\", usecols=[0,1], \\\n",
    "                                  names=['chrom', 'size'], na_filter=False, dtype={'chrom':'str', 'size':'int'})\n",
    "        chr_list = list(chr_df['chrom'])\n",
    "        CHROMOSOMES = [word for word in chr_list if len(word) <= 6]\n",
    "    else:\n",
    "        raise ValueError(\"Chromosome size file missing from generated_sequences(sample_simulated.chrom.sizes)!\") \n",
    "\n",
    "    windowbed.seek(0)\n",
    "    windowbed_csv_reader = csv.reader(windowbed, delimiter='\\t')\n",
    "    windowbed_line = next(windowbed_csv_reader)\n",
    "    while(windowbed_line[0][0] == '#'):\n",
    "        windowbed_line = next(windowbed_csv_reader)    \n",
    "\n",
    "    motif_stats = []\n",
    "    sorted_motif_stats = []\n",
    "    tf_motif_path = outdir + '/motifs/simulated/*'\n",
    "    motif_filenames = glob.glob(tf_motif_path)\n",
    "    motif_count = len(motif_filenames)\n",
    "    print(\"Processing %s motif files in %s\" % motif_count, tf_motif_path)\n",
    "    for filename in motif_filenames:\n",
    "        filename_no_path = filename.split('/')[-1]\n",
    "        if os.path.getsize(filename) > 0 and os.path.basename(filename).endswith(tuple(['.bed'])):\n",
    "            [md_score, small_window, large_window, motif_site_count, heat] = get_md_score(filename, int(threads), windowbed_name, CHROMOSOMES, window)\n",
    "            print('The MD-score for the provided BED regions vs %s is %.6f' % (filename_no_path, md_score))\n",
    "            motif_stats.append({ 'motif_file': filename_no_path,'md_score': md_score, 'small_window': small_window, 'large_window': large_window, 'motif_site_count': motif_site_count, 'heat': heat })\n",
    "    # sort the stats dictionary by MD-score, descending order\n",
    "    sorted_motif_stats = sorted(motif_stats, key=itemgetter('md_score'), reverse=True)\n",
    "\n",
    "    md_score_fp = open(\"%s/final_output/%s_simulated_md_scores.txt\" % (outdir, sample), 'w')\n",
    "    for stat in sorted_motif_stats:\n",
    "        md_score_fp.write(\"%s,%s,%s,%s,%s,%s\\n\" % (stat['motif_file'], stat['md_score'], stat['small_window'], \\\n",
    "                           stat['large_window'], stat['motif_site_count'], stat['heat']))\n",
    "    md_score_fp.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mdcalc_prescan(annotation, window, threads, outdir, sample, genome, pre_scan):\n",
    "    windowbed_name = annotation \n",
    "    input_file_statinfo = os.stat(windowbed_name)    \n",
    "    if input_file_statinfo.st_size !=0:\n",
    "        windowbed = open(windowbed_name)\n",
    "        windowbed_csv_reader = csv.reader(windowbed, delimiter='\\t')\n",
    "        windowbed_line = next(windowbed_csv_reader)\n",
    "        windowbed_peak_count = 0        \n",
    "    else:\n",
    "        raise ValueError(\"The annotation is empty!\")\n",
    "\n",
    "        \n",
    "    g = genome.split('/')[-1].split('.')[-2]    \n",
    "    chromosomes = outdir + \"/generated_sequences/\" + g + \".chrom.sizes\"\n",
    "    if chromosomes:\n",
    "        chr_df = pd.read_csv(chromosomes, header=None, comment='#', sep=\"\\t\", usecols=[0,1], names=['chrom', 'size'], na_filter=False, dtype={'chrom':'str', 'size':'int'})\n",
    "        chr_list = list(chr_df['chrom'])\n",
    "        CHROMOSOMES = [word for word in chr_list if len(word) <= 6]\n",
    "    else:\n",
    "        raise ValueError(\"Chromosome size file missing from generated_sequences(genome.chrom.sizes)!\") \n",
    "\n",
    "    windowbed.seek(0)\n",
    "    windowbed_csv_reader = csv.reader(windowbed, delimiter='\\t')\n",
    "    windowbed_line = next(windowbed_csv_reader)\n",
    "    while(windowbed_line[0][0] == '#'):\n",
    "        windowbed_line = next(windowbed_csv_reader)    \n",
    "\n",
    "    motif_stats = []\n",
    "    sorted_motif_stats = []\n",
    "    tf_motif_path = pre_scan + '/*'\n",
    "    motif_filenames = glob.glob(tf_motif_path)\n",
    "    motif_count = len(motif_filenames)\n",
    "    print(\"Processing %s motif files in %s\" % motif_count, tf_motif_path)\n",
    "    for filename in motif_filenames:\n",
    "        filename_no_path = filename.split('/')[-1]\n",
    "        if os.path.getsize(filename) > 0 and os.path.basename(filename).endswith(tuple(['.bed'])):\n",
    "            [md_score, small_window, large_window, motif_site_count, heat] = get_md_score(filename, int(threads), windowbed_name, CHROMOSOMES, window)\n",
    "            print('The MD-score for the provided BED regions vs %s is %.6f' % (filename_no_path, md_score))\n",
    "            motif_stats.append({ 'motif_file': filename_no_path,'md_score': md_score, 'small_window': small_window, 'large_window': large_window, 'motif_site_count': motif_site_count, 'heat': heat })\n",
    "    # sort the stats dictionary by MD-score, descending order\n",
    "    sorted_motif_stats = sorted(motif_stats, key=itemgetter('md_score'), reverse=True)\n",
    "\n",
    "    md_score_fp = open(\"%s/final_output/%s_experimental_md_scores.txt\" % (outdir, sample), 'w')\n",
    "    for stat in sorted_motif_stats:\n",
    "        md_score_fp.write(\"%s,%s,%s,%s,%s,%s\\n\" % (stat['motif_file'], stat['md_score'], stat['small_window'], stat['large_window'], stat['motif_site_count'], stat['heat']))\n",
    "    md_score_fp.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Additional MD Code for prescan\n",
    "window, genome for args\n",
    "if os.path.getsize(filename) > 0 and os.path.basename(filename).endswith(tuple(['.bed', '.BedGraph', '.txt']\n",
    "##do this for the -p flag !\n",
    "    if not os.listdir(motifs):\n",
    "        raise NameError(\"The specified motif directory is empty or does not exist. Please check that there are motif files in the directory specified by the -p/--pre_scan argument.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "4412\n",
      "13328954\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "##def expt_bed and sim_bed need to be fixed.\n",
    "# '''This function takes the center of all sequences and generates a bedfile for dowindownstream use\n",
    "# '''\n",
    "outdir = '/scratch/Users/tajo5912/rbg/test'\n",
    "sample = 'test'\n",
    "window = 1500\n",
    "chrom_num=10\n",
    "\n",
    "\n",
    "df = pd.read_csv(outdir + \"/generated_sequences/\" + sample + \"_experimental_window_sequences.fa\", header=None)\n",
    "df = df[df.index % 2 != 0]\n",
    "\n",
    "l_expt_seq = df.values.tolist()\n",
    "flat_list = [item for sublist in l_expt_seq for item in sublist]\n",
    "strN = str('N')*20\n",
    "file = [\"{}{}\".format(i,strN) for i in flat_list]\n",
    "\n",
    "sequence_num = len(file)\n",
    "seq_length = ((2*int(window))+21)\n",
    "tot_bases = int(sequence_num*seq_length)\n",
    "seq_per_chrom = math.floor(sequence_num/chrom_num)\n",
    "remainder_seq_per_chrom = (sequence_num%chrom_num)\n",
    "print(remainder_seq_per_chrom)\n",
    "print(seq_per_chrom)\n",
    "base_per_chrom = int(tot_bases/chrom_num) \n",
    "print(base_per_chrom)\n",
    "\n",
    "# dbed = {}\n",
    "# for i in range(0,(int(seq_per_chrom))):\n",
    "#         dbed['r_' + str(i)] = [int(window-1)+i*(2*window+21), int(window+1)+i*(2*window+21)]\n",
    "# dfbed = pd.DataFrame.from_dict(dbed, orient='index')\n",
    "# dfbed.columns = ['start', 'stop']\n",
    "# dfbed = dfbed.reset_index()\n",
    "# dfbed = pd.concat([dfbed]*chrom_num)\n",
    "# dfbed\n",
    "# l_expt=[]\n",
    "# for i in range(int(seq_per_chrom)):\n",
    "#     for c in range(int(chrom_num)):\n",
    "#         l_expt.append(str('exp_' + str(c+1)))\n",
    "# l_expt = sorted(l_expt)\n",
    "# len(l_expt)\n",
    "# dfbed['exp'] = l_expt\n",
    "# dfbed = dfbed[['exp', 'start', 'stop']]\n",
    "# dfbed\n",
    "# #print(chrom_num)\n",
    "# #dfbed.to_csv(outdir + \"/annotations/\" + str(sample) + \"_experimental_genome_centered.bed\", header=None, index=False, sep='\\t')\n",
    "\n",
    "# # #creating the windowed bed\n",
    "\n",
    "# dfbed[\"start_new\"] = dfbed.apply(lambda x: round((x[\"start\"] + x[\"stop\"])/2), axis=1)\n",
    "# dfbed[\"stop_new\"] = dfbed.apply(lambda x: x[\"start_new\"] + 1, axis = 1)\n",
    "\n",
    "# ##the -1500 position from \"origin\"\n",
    "# dfbed[\"start\"] = dfbed.apply(lambda x: x[\"start_new\"] - int(window), axis=1)\n",
    "\n",
    "# ##the 1500 position from the \"origin\"\n",
    "# dfbed[\"stop\"] = dfbed.apply(lambda x: x[\"stop_new\"] + int(window), axis=1)\n",
    "\n",
    "# ##saving the new annotation\n",
    "# dfbed.to_csv(outdir + '/annotations/' + sample + '_experimental_genome_window.bed', sep='\\t', columns=[\"exp\",\"start\",\"stop\"], header = False, index = False)\n",
    "# print('Experimental Bed Files Genereated: ' + outdir + \"/annotations/\" + str(sample) + \"_experimental_genome_centered.bed and \" + str(sample) + \"_experimental_genome_window.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###SIMULATED\n",
    "'''This function takes the center of all simulated sequences and generates a bedfile for dowindownstream use\n",
    "'''\n",
    "seq_per_chrom = int(self.sequence_num/self.chrom_num)\n",
    "dbed = {}\n",
    "for i in range(0,(int(seq_per_chrom))):\n",
    "        dbed['r_' + str(i)] = [int(self.window-1)+i*(2*self.window+21), int(self.window+1)+i*(2*self.window+21)]\n",
    "dfbed = pd.DataFrame.from_dict(dbed, orient='index')\n",
    "dfbed.columns = ['start', 'stop']\n",
    "dfbed = dfbed.reset_index()\n",
    "dfbed = pd.concat([dfbed]*self.chrom_num)\n",
    "\n",
    "l_sim=[]\n",
    "for i in range(int(seq_per_chrom)):\n",
    "    for c in range(int(self.chrom_num)):\n",
    "        l_sim.append(str('sim_' + str(c+1)))\n",
    "l_sim = sorted(l_sim)\n",
    "dfbed['sim'] = l_sim\n",
    "dfbed = dfbed[['sim', 'start', 'stop']]\n",
    "dfbed.to_csv(self.outdir + \"/annotations/\" + str(self.sample) + \"_simulated_centered.bed\", header=None, index=False, sep='\\t')\n",
    "\n",
    "#creating the windowed bed\n",
    "\n",
    "dfbed[\"start_new\"] = dfbed.apply(lambda x: round((x[\"start\"] + x[\"stop\"])/2), axis=1)\n",
    "dfbed[\"stop_new\"] = dfbed.apply(lambda x: x[\"start_new\"] + 1, axis = 1)\n",
    "\n",
    "##the -1500 position from \"origin\"\n",
    "dfbed[\"start\"] = dfbed.apply(lambda x: x[\"start_new\"] - int(self.window), axis=1)\n",
    "\n",
    "##the 1500 position from the \"origin\"\n",
    "dfbed[\"stop\"] = dfbed.apply(lambda x: x[\"stop_new\"] + int(self.window), axis=1)\n",
    "\n",
    "##saving the new annotation\n",
    "dfbed.to_csv(self.outdir + '/annotations/' + self.sample + '_simulated_window.bed', sep='\\t', columns=[\"sim\",\"start\",\"stop\"], header = False, index = False)\n",
    "return print('Simulated Bed Files Genereated: ' + self.outdir + \"/annotations/\" + str(self.sample) + \"_simulated_centered.bed and \" + str(self.sample) + \"_simulated_window.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Post-processing FIMO output for TP53_M06704_1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>identifier</th>\n",
       "      <th>motif_id</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>matched_sequence</th>\n",
       "      <th>percent_gc</th>\n",
       "      <th>mean_percent_gc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>1607701</td>\n",
       "      <td>1607720</td>\n",
       "      <td>sim_1:1607701.0-1607720.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.23</td>\n",
       "      <td>+</td>\n",
       "      <td>AGGCTTGCAAGGACATGCCT</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>12024196</td>\n",
       "      <td>12024215</td>\n",
       "      <td>sim_1:12024196.0-12024215.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>19.33</td>\n",
       "      <td>+</td>\n",
       "      <td>CGCCATGCCGGGACATGCCT</td>\n",
       "      <td>0.70</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>2943101</td>\n",
       "      <td>2943120</td>\n",
       "      <td>sim_1:2943101.0-2943120.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.14</td>\n",
       "      <td>-</td>\n",
       "      <td>AAACTTGCTCAGACATGACT</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>5926733</td>\n",
       "      <td>5926752</td>\n",
       "      <td>sim_1:5926733.0-5926752.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.90</td>\n",
       "      <td>-</td>\n",
       "      <td>CAAGATGCTCAGACATGTTT</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>4174561</td>\n",
       "      <td>4174580</td>\n",
       "      <td>sim_1:4174561.0-4174580.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>19.17</td>\n",
       "      <td>+</td>\n",
       "      <td>TTATATGCTCAGACATGTCT</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>8189020</td>\n",
       "      <td>8189039</td>\n",
       "      <td>sim_1:8189020.0-8189039.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.28</td>\n",
       "      <td>+</td>\n",
       "      <td>ATACAAGGTCAGGCATGCCC</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>11055590</td>\n",
       "      <td>11055609</td>\n",
       "      <td>sim_1:11055590.0-11055609.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.31</td>\n",
       "      <td>+</td>\n",
       "      <td>CGACATGTATAGACTTGTTC</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>5063863</td>\n",
       "      <td>5063882</td>\n",
       "      <td>sim_1:5063863.0-5063882.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>19.82</td>\n",
       "      <td>+</td>\n",
       "      <td>GTACTTGTCCAAGCATGTCC</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>7230857</td>\n",
       "      <td>7230876</td>\n",
       "      <td>sim_1:7230857.0-7230876.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>18.80</td>\n",
       "      <td>-</td>\n",
       "      <td>AGACTTGTTTGAGCATGTCC</td>\n",
       "      <td>0.45</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>148944</td>\n",
       "      <td>148963</td>\n",
       "      <td>sim_1:148944.0-148963.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>20.00</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCCATGCTGGGACATGTCC</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>11822083</td>\n",
       "      <td>11822102</td>\n",
       "      <td>sim_1:11822083.0-11822102.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.94</td>\n",
       "      <td>+</td>\n",
       "      <td>AGGCAAGCCTGAGCATGCCC</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>11822083</td>\n",
       "      <td>11822102</td>\n",
       "      <td>sim_1:11822083.0-11822102.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>20.29</td>\n",
       "      <td>-</td>\n",
       "      <td>GGGCATGCTCAGGCTTGCCT</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>5105820</td>\n",
       "      <td>5105839</td>\n",
       "      <td>sim_1:5105820.0-5105839.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.02</td>\n",
       "      <td>-</td>\n",
       "      <td>AGACTTGACCGGACTTGCTT</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>7958195</td>\n",
       "      <td>7958214</td>\n",
       "      <td>sim_1:7958195.0-7958214.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.05</td>\n",
       "      <td>+</td>\n",
       "      <td>GATCATGCCAGGACCTGCCT</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>5937536</td>\n",
       "      <td>5937555</td>\n",
       "      <td>sim_1:5937536.0-5937555.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.69</td>\n",
       "      <td>+</td>\n",
       "      <td>CCTCATGCCCGGGCCTGTCC</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>7626888</td>\n",
       "      <td>7626907</td>\n",
       "      <td>sim_1:7626888.0-7626907.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.49</td>\n",
       "      <td>+</td>\n",
       "      <td>AAGCACGCCGAGGCATGCCT</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>7230857</td>\n",
       "      <td>7230876</td>\n",
       "      <td>sim_1:7230857.0-7230876.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>18.41</td>\n",
       "      <td>+</td>\n",
       "      <td>GGACATGCTCAAACAAGTCT</td>\n",
       "      <td>0.45</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>3815569</td>\n",
       "      <td>3815588</td>\n",
       "      <td>sim_1:3815569.0-3815588.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>16.81</td>\n",
       "      <td>-</td>\n",
       "      <td>ATACATGCCGTGGCTTGCTT</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>7329820</td>\n",
       "      <td>7329839</td>\n",
       "      <td>sim_1:7329820.0-7329839.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.85</td>\n",
       "      <td>+</td>\n",
       "      <td>GAACATGCAGAGACCTGTCC</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>sim_1</td>\n",
       "      <td>5063863</td>\n",
       "      <td>5063882</td>\n",
       "      <td>sim_1:5063863.0-5063882.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.57</td>\n",
       "      <td>-</td>\n",
       "      <td>GGACATGCTTGGACAAGTAC</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>sim_10</td>\n",
       "      <td>9959648</td>\n",
       "      <td>9959667</td>\n",
       "      <td>sim_10:9959648.0-9959667.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.21</td>\n",
       "      <td>-</td>\n",
       "      <td>CCACATGTTCAAACTTGTCC</td>\n",
       "      <td>0.45</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>sim_10</td>\n",
       "      <td>4992928</td>\n",
       "      <td>4992947</td>\n",
       "      <td>sim_10:4992928.0-4992947.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.78</td>\n",
       "      <td>+</td>\n",
       "      <td>AAACATGTACAAACATGCCA</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>sim_10</td>\n",
       "      <td>1292466</td>\n",
       "      <td>1292485</td>\n",
       "      <td>sim_10:1292466.0-1292485.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.27</td>\n",
       "      <td>-</td>\n",
       "      <td>ACACATGTTTACGCATGTCC</td>\n",
       "      <td>0.45</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>sim_10</td>\n",
       "      <td>7547595</td>\n",
       "      <td>7547614</td>\n",
       "      <td>sim_10:7547595.0-7547614.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.87</td>\n",
       "      <td>-</td>\n",
       "      <td>GTAGATGTACAGACATGTCT</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>sim_10</td>\n",
       "      <td>508495</td>\n",
       "      <td>508514</td>\n",
       "      <td>sim_10:508495.0-508514.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.28</td>\n",
       "      <td>+</td>\n",
       "      <td>TTATATGTCTTGGCATGTCC</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>sim_10</td>\n",
       "      <td>569112</td>\n",
       "      <td>569131</td>\n",
       "      <td>sim_10:569112.0-569131.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>18.63</td>\n",
       "      <td>+</td>\n",
       "      <td>AGACATGCCCGGACAAGTGT</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>sim_10</td>\n",
       "      <td>12749689</td>\n",
       "      <td>12749708</td>\n",
       "      <td>sim_10:12749689.0-12749708.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>18.86</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCATGTTAAGGCATGTCT</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>sim_10</td>\n",
       "      <td>1756207</td>\n",
       "      <td>1756226</td>\n",
       "      <td>sim_10:1756207.0-1756226.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.89</td>\n",
       "      <td>-</td>\n",
       "      <td>CGACATGTCCGGGCATGGTT</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>sim_10</td>\n",
       "      <td>10940555</td>\n",
       "      <td>10940574</td>\n",
       "      <td>sim_10:10940555.0-10940574.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>18.54</td>\n",
       "      <td>+</td>\n",
       "      <td>AGGCATGCCCGGGCTCGTCC</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>sim_10</td>\n",
       "      <td>2024958</td>\n",
       "      <td>2024977</td>\n",
       "      <td>sim_10:2024958.0-2024977.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>18.79</td>\n",
       "      <td>+</td>\n",
       "      <td>AAAGATGCCCGTACATGCCT</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>sim_7</td>\n",
       "      <td>3534977</td>\n",
       "      <td>3534996</td>\n",
       "      <td>sim_7:3534977.0-3534996.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>18.08</td>\n",
       "      <td>-</td>\n",
       "      <td>ATACCTGTCCGGACTTGTTT</td>\n",
       "      <td>0.45</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>sim_7</td>\n",
       "      <td>4345303</td>\n",
       "      <td>4345322</td>\n",
       "      <td>sim_7:4345303.0-4345322.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.36</td>\n",
       "      <td>+</td>\n",
       "      <td>GAACTTGTCCGAACTTGCTC</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>sim_7</td>\n",
       "      <td>7204928</td>\n",
       "      <td>7204947</td>\n",
       "      <td>sim_7:7204928.0-7204947.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.12</td>\n",
       "      <td>+</td>\n",
       "      <td>GGAAATGTCCAGACCTGTCT</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>sim_7</td>\n",
       "      <td>5444588</td>\n",
       "      <td>5444607</td>\n",
       "      <td>sim_7:5444588.0-5444607.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>20.40</td>\n",
       "      <td>+</td>\n",
       "      <td>GGACATGTCGGCACATGTCC</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>sim_8</td>\n",
       "      <td>8240230</td>\n",
       "      <td>8240249</td>\n",
       "      <td>sim_8:8240230.0-8240249.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>16.98</td>\n",
       "      <td>-</td>\n",
       "      <td>GAACTTGTCGTGACAAGTCC</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>sim_8</td>\n",
       "      <td>2863179</td>\n",
       "      <td>2863198</td>\n",
       "      <td>sim_8:2863179.0-2863198.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.89</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGCATGCTCGGACATGCCA</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>sim_8</td>\n",
       "      <td>9531943</td>\n",
       "      <td>9531962</td>\n",
       "      <td>sim_8:9531943.0-9531962.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.42</td>\n",
       "      <td>-</td>\n",
       "      <td>GCAGATGCCAAGACATGCCT</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>sim_8</td>\n",
       "      <td>12593060</td>\n",
       "      <td>12593079</td>\n",
       "      <td>sim_8:12593060.0-12593079.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.84</td>\n",
       "      <td>-</td>\n",
       "      <td>GGACACGCCTAAACATGTCC</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>sim_8</td>\n",
       "      <td>6929060</td>\n",
       "      <td>6929079</td>\n",
       "      <td>sim_8:6929060.0-6929079.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.09</td>\n",
       "      <td>-</td>\n",
       "      <td>TGACATGTTCACACTTGTCT</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>sim_8</td>\n",
       "      <td>3363158</td>\n",
       "      <td>3363177</td>\n",
       "      <td>sim_8:3363158.0-3363177.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>16.82</td>\n",
       "      <td>-</td>\n",
       "      <td>GAACAAGCTCATGCTTGCCC</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>sim_8</td>\n",
       "      <td>9229583</td>\n",
       "      <td>9229602</td>\n",
       "      <td>sim_8:9229583.0-9229602.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>18.45</td>\n",
       "      <td>-</td>\n",
       "      <td>ACACATGCGGGGACATGCCT</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>sim_8</td>\n",
       "      <td>12002351</td>\n",
       "      <td>12002370</td>\n",
       "      <td>sim_8:12002351.0-12002370.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>18.20</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTTATGTCCAGGCATGCTC</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>sim_8</td>\n",
       "      <td>3434788</td>\n",
       "      <td>3434807</td>\n",
       "      <td>sim_8:3434788.0-3434807.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>18.45</td>\n",
       "      <td>+</td>\n",
       "      <td>AAACTCGCCCGGACATGTTT</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>sim_8</td>\n",
       "      <td>13006408</td>\n",
       "      <td>13006427</td>\n",
       "      <td>sim_8:13006408.0-13006427.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.49</td>\n",
       "      <td>-</td>\n",
       "      <td>AGCCATGCTGAGACATGCTC</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>sim_8</td>\n",
       "      <td>6879643</td>\n",
       "      <td>6879662</td>\n",
       "      <td>sim_8:6879643.0-6879662.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>18.01</td>\n",
       "      <td>-</td>\n",
       "      <td>CAACATGCATGGGCATGCAC</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>sim_8</td>\n",
       "      <td>11380762</td>\n",
       "      <td>11380781</td>\n",
       "      <td>sim_8:11380762.0-11380781.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.03</td>\n",
       "      <td>-</td>\n",
       "      <td>AATAATGTCCAGACATGTTT</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>sim_8</td>\n",
       "      <td>8334661</td>\n",
       "      <td>8334680</td>\n",
       "      <td>sim_8:8334661.0-8334680.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>19.48</td>\n",
       "      <td>-</td>\n",
       "      <td>ATACATGCCTAGGCCTGCCC</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>sim_9</td>\n",
       "      <td>2011119</td>\n",
       "      <td>2011138</td>\n",
       "      <td>sim_9:2011119.0-2011138.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>16.81</td>\n",
       "      <td>-</td>\n",
       "      <td>AATCATGCCCGAGCACGTCC</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>sim_9</td>\n",
       "      <td>7275197</td>\n",
       "      <td>7275216</td>\n",
       "      <td>sim_9:7275197.0-7275216.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>18.17</td>\n",
       "      <td>-</td>\n",
       "      <td>GAACTTGCACAAACATGTCT</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>sim_9</td>\n",
       "      <td>10958483</td>\n",
       "      <td>10958502</td>\n",
       "      <td>sim_9:10958483.0-10958502.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.10</td>\n",
       "      <td>+</td>\n",
       "      <td>GAGCATGTCTGTGCTTGCCC</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>sim_9</td>\n",
       "      <td>9658299</td>\n",
       "      <td>9658318</td>\n",
       "      <td>sim_9:9658299.0-9658318.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.10</td>\n",
       "      <td>+</td>\n",
       "      <td>ATACAAGCGCTGGCATGCCC</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>sim_9</td>\n",
       "      <td>12403194</td>\n",
       "      <td>12403213</td>\n",
       "      <td>sim_9:12403194.0-12403213.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.66</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGGATGTTCAGGCATGCCC</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>sim_9</td>\n",
       "      <td>5905481</td>\n",
       "      <td>5905500</td>\n",
       "      <td>sim_9:5905481.0-5905500.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.64</td>\n",
       "      <td>-</td>\n",
       "      <td>GCGTATGTCGGGACATGTCC</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>sim_9</td>\n",
       "      <td>2656965</td>\n",
       "      <td>2656984</td>\n",
       "      <td>sim_9:2656965.0-2656984.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.30</td>\n",
       "      <td>-</td>\n",
       "      <td>GGATATGTTGAGACATGCTC</td>\n",
       "      <td>0.45</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>sim_9</td>\n",
       "      <td>80158</td>\n",
       "      <td>80177</td>\n",
       "      <td>sim_9:80158.0-80177.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.42</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCATGCACTGACATGTTC</td>\n",
       "      <td>0.45</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>sim_9</td>\n",
       "      <td>10958483</td>\n",
       "      <td>10958502</td>\n",
       "      <td>sim_9:10958483.0-10958502.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.38</td>\n",
       "      <td>-</td>\n",
       "      <td>GGGCAAGCACAGACATGCTC</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>sim_9</td>\n",
       "      <td>3765592</td>\n",
       "      <td>3765611</td>\n",
       "      <td>sim_9:3765592.0-3765611.0_-;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.38</td>\n",
       "      <td>-</td>\n",
       "      <td>CAATATGTCCAGACCTGTCT</td>\n",
       "      <td>0.45</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>sim_9</td>\n",
       "      <td>8657229</td>\n",
       "      <td>8657248</td>\n",
       "      <td>sim_9:8657229.0-8657248.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>17.40</td>\n",
       "      <td>+</td>\n",
       "      <td>CGACCTGTCGGGACATGCCA</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>sim_9</td>\n",
       "      <td>1867668</td>\n",
       "      <td>1867687</td>\n",
       "      <td>sim_9:1867668.0-1867687.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>20.43</td>\n",
       "      <td>+</td>\n",
       "      <td>GAATATGTCCTGGCATGTCT</td>\n",
       "      <td>0.45</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>sim_9</td>\n",
       "      <td>6959969</td>\n",
       "      <td>6959988</td>\n",
       "      <td>sim_9:6959969.0-6959988.0_+;TP53_M06704_1</td>\n",
       "      <td>TP53_M06704_1</td>\n",
       "      <td>18.85</td>\n",
       "      <td>+</td>\n",
       "      <td>CTACATGCTCACACATGTCC</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.533894</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>208 rows  10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sequence_name     start      stop  \\\n",
       "0           sim_1   1607701   1607720   \n",
       "1           sim_1  12024196  12024215   \n",
       "2           sim_1   2943101   2943120   \n",
       "3           sim_1   5926733   5926752   \n",
       "4           sim_1   4174561   4174580   \n",
       "5           sim_1   8189020   8189039   \n",
       "6           sim_1  11055590  11055609   \n",
       "7           sim_1   5063863   5063882   \n",
       "8           sim_1   7230857   7230876   \n",
       "9           sim_1    148944    148963   \n",
       "10          sim_1  11822083  11822102   \n",
       "11          sim_1  11822083  11822102   \n",
       "12          sim_1   5105820   5105839   \n",
       "13          sim_1   7958195   7958214   \n",
       "14          sim_1   5937536   5937555   \n",
       "15          sim_1   7626888   7626907   \n",
       "16          sim_1   7230857   7230876   \n",
       "17          sim_1   3815569   3815588   \n",
       "18          sim_1   7329820   7329839   \n",
       "19          sim_1   5063863   5063882   \n",
       "20         sim_10   9959648   9959667   \n",
       "21         sim_10   4992928   4992947   \n",
       "22         sim_10   1292466   1292485   \n",
       "23         sim_10   7547595   7547614   \n",
       "24         sim_10    508495    508514   \n",
       "25         sim_10    569112    569131   \n",
       "26         sim_10  12749689  12749708   \n",
       "27         sim_10   1756207   1756226   \n",
       "28         sim_10  10940555  10940574   \n",
       "29         sim_10   2024958   2024977   \n",
       "..            ...       ...       ...   \n",
       "178         sim_7   3534977   3534996   \n",
       "179         sim_7   4345303   4345322   \n",
       "180         sim_7   7204928   7204947   \n",
       "181         sim_7   5444588   5444607   \n",
       "182         sim_8   8240230   8240249   \n",
       "183         sim_8   2863179   2863198   \n",
       "184         sim_8   9531943   9531962   \n",
       "185         sim_8  12593060  12593079   \n",
       "186         sim_8   6929060   6929079   \n",
       "187         sim_8   3363158   3363177   \n",
       "188         sim_8   9229583   9229602   \n",
       "189         sim_8  12002351  12002370   \n",
       "190         sim_8   3434788   3434807   \n",
       "191         sim_8  13006408  13006427   \n",
       "192         sim_8   6879643   6879662   \n",
       "193         sim_8  11380762  11380781   \n",
       "194         sim_8   8334661   8334680   \n",
       "195         sim_9   2011119   2011138   \n",
       "196         sim_9   7275197   7275216   \n",
       "197         sim_9  10958483  10958502   \n",
       "198         sim_9   9658299   9658318   \n",
       "199         sim_9  12403194  12403213   \n",
       "200         sim_9   5905481   5905500   \n",
       "201         sim_9   2656965   2656984   \n",
       "202         sim_9     80158     80177   \n",
       "203         sim_9  10958483  10958502   \n",
       "204         sim_9   3765592   3765611   \n",
       "205         sim_9   8657229   8657248   \n",
       "206         sim_9   1867668   1867687   \n",
       "207         sim_9   6959969   6959988   \n",
       "\n",
       "                                       identifier       motif_id  score  \\\n",
       "0       sim_1:1607701.0-1607720.0_+;TP53_M06704_1  TP53_M06704_1  17.23   \n",
       "1     sim_1:12024196.0-12024215.0_+;TP53_M06704_1  TP53_M06704_1  19.33   \n",
       "2       sim_1:2943101.0-2943120.0_-;TP53_M06704_1  TP53_M06704_1  17.14   \n",
       "3       sim_1:5926733.0-5926752.0_-;TP53_M06704_1  TP53_M06704_1  17.90   \n",
       "4       sim_1:4174561.0-4174580.0_+;TP53_M06704_1  TP53_M06704_1  19.17   \n",
       "5       sim_1:8189020.0-8189039.0_+;TP53_M06704_1  TP53_M06704_1  17.28   \n",
       "6     sim_1:11055590.0-11055609.0_+;TP53_M06704_1  TP53_M06704_1  17.31   \n",
       "7       sim_1:5063863.0-5063882.0_+;TP53_M06704_1  TP53_M06704_1  19.82   \n",
       "8       sim_1:7230857.0-7230876.0_-;TP53_M06704_1  TP53_M06704_1  18.80   \n",
       "9         sim_1:148944.0-148963.0_-;TP53_M06704_1  TP53_M06704_1  20.00   \n",
       "10    sim_1:11822083.0-11822102.0_+;TP53_M06704_1  TP53_M06704_1  17.94   \n",
       "11    sim_1:11822083.0-11822102.0_-;TP53_M06704_1  TP53_M06704_1  20.29   \n",
       "12      sim_1:5105820.0-5105839.0_-;TP53_M06704_1  TP53_M06704_1  17.02   \n",
       "13      sim_1:7958195.0-7958214.0_+;TP53_M06704_1  TP53_M06704_1  17.05   \n",
       "14      sim_1:5937536.0-5937555.0_+;TP53_M06704_1  TP53_M06704_1  17.69   \n",
       "15      sim_1:7626888.0-7626907.0_+;TP53_M06704_1  TP53_M06704_1  17.49   \n",
       "16      sim_1:7230857.0-7230876.0_+;TP53_M06704_1  TP53_M06704_1  18.41   \n",
       "17      sim_1:3815569.0-3815588.0_-;TP53_M06704_1  TP53_M06704_1  16.81   \n",
       "18      sim_1:7329820.0-7329839.0_+;TP53_M06704_1  TP53_M06704_1  17.85   \n",
       "19      sim_1:5063863.0-5063882.0_-;TP53_M06704_1  TP53_M06704_1  17.57   \n",
       "20     sim_10:9959648.0-9959667.0_-;TP53_M06704_1  TP53_M06704_1  17.21   \n",
       "21     sim_10:4992928.0-4992947.0_+;TP53_M06704_1  TP53_M06704_1  17.78   \n",
       "22     sim_10:1292466.0-1292485.0_-;TP53_M06704_1  TP53_M06704_1  17.27   \n",
       "23     sim_10:7547595.0-7547614.0_-;TP53_M06704_1  TP53_M06704_1  17.87   \n",
       "24       sim_10:508495.0-508514.0_+;TP53_M06704_1  TP53_M06704_1  17.28   \n",
       "25       sim_10:569112.0-569131.0_+;TP53_M06704_1  TP53_M06704_1  18.63   \n",
       "26   sim_10:12749689.0-12749708.0_+;TP53_M06704_1  TP53_M06704_1  18.86   \n",
       "27     sim_10:1756207.0-1756226.0_-;TP53_M06704_1  TP53_M06704_1  17.89   \n",
       "28   sim_10:10940555.0-10940574.0_+;TP53_M06704_1  TP53_M06704_1  18.54   \n",
       "29     sim_10:2024958.0-2024977.0_+;TP53_M06704_1  TP53_M06704_1  18.79   \n",
       "..                                            ...            ...    ...   \n",
       "178     sim_7:3534977.0-3534996.0_-;TP53_M06704_1  TP53_M06704_1  18.08   \n",
       "179     sim_7:4345303.0-4345322.0_+;TP53_M06704_1  TP53_M06704_1  17.36   \n",
       "180     sim_7:7204928.0-7204947.0_+;TP53_M06704_1  TP53_M06704_1  17.12   \n",
       "181     sim_7:5444588.0-5444607.0_+;TP53_M06704_1  TP53_M06704_1  20.40   \n",
       "182     sim_8:8240230.0-8240249.0_-;TP53_M06704_1  TP53_M06704_1  16.98   \n",
       "183     sim_8:2863179.0-2863198.0_-;TP53_M06704_1  TP53_M06704_1  17.89   \n",
       "184     sim_8:9531943.0-9531962.0_-;TP53_M06704_1  TP53_M06704_1  17.42   \n",
       "185   sim_8:12593060.0-12593079.0_-;TP53_M06704_1  TP53_M06704_1  17.84   \n",
       "186     sim_8:6929060.0-6929079.0_-;TP53_M06704_1  TP53_M06704_1  17.09   \n",
       "187     sim_8:3363158.0-3363177.0_-;TP53_M06704_1  TP53_M06704_1  16.82   \n",
       "188     sim_8:9229583.0-9229602.0_-;TP53_M06704_1  TP53_M06704_1  18.45   \n",
       "189   sim_8:12002351.0-12002370.0_-;TP53_M06704_1  TP53_M06704_1  18.20   \n",
       "190     sim_8:3434788.0-3434807.0_+;TP53_M06704_1  TP53_M06704_1  18.45   \n",
       "191   sim_8:13006408.0-13006427.0_-;TP53_M06704_1  TP53_M06704_1  17.49   \n",
       "192     sim_8:6879643.0-6879662.0_-;TP53_M06704_1  TP53_M06704_1  18.01   \n",
       "193   sim_8:11380762.0-11380781.0_-;TP53_M06704_1  TP53_M06704_1  17.03   \n",
       "194     sim_8:8334661.0-8334680.0_-;TP53_M06704_1  TP53_M06704_1  19.48   \n",
       "195     sim_9:2011119.0-2011138.0_-;TP53_M06704_1  TP53_M06704_1  16.81   \n",
       "196     sim_9:7275197.0-7275216.0_-;TP53_M06704_1  TP53_M06704_1  18.17   \n",
       "197   sim_9:10958483.0-10958502.0_+;TP53_M06704_1  TP53_M06704_1  17.10   \n",
       "198     sim_9:9658299.0-9658318.0_+;TP53_M06704_1  TP53_M06704_1  17.10   \n",
       "199   sim_9:12403194.0-12403213.0_-;TP53_M06704_1  TP53_M06704_1  17.66   \n",
       "200     sim_9:5905481.0-5905500.0_-;TP53_M06704_1  TP53_M06704_1  17.64   \n",
       "201     sim_9:2656965.0-2656984.0_-;TP53_M06704_1  TP53_M06704_1  17.30   \n",
       "202         sim_9:80158.0-80177.0_+;TP53_M06704_1  TP53_M06704_1  17.42   \n",
       "203   sim_9:10958483.0-10958502.0_-;TP53_M06704_1  TP53_M06704_1  17.38   \n",
       "204     sim_9:3765592.0-3765611.0_-;TP53_M06704_1  TP53_M06704_1  17.38   \n",
       "205     sim_9:8657229.0-8657248.0_+;TP53_M06704_1  TP53_M06704_1  17.40   \n",
       "206     sim_9:1867668.0-1867687.0_+;TP53_M06704_1  TP53_M06704_1  20.43   \n",
       "207     sim_9:6959969.0-6959988.0_+;TP53_M06704_1  TP53_M06704_1  18.85   \n",
       "\n",
       "    strand      matched_sequence  percent_gc  mean_percent_gc  \n",
       "0        +  AGGCTTGCAAGGACATGCCT        0.55         0.533894  \n",
       "1        +  CGCCATGCCGGGACATGCCT        0.70         0.533894  \n",
       "2        -  AAACTTGCTCAGACATGACT        0.40         0.533894  \n",
       "3        -  CAAGATGCTCAGACATGTTT        0.40         0.533894  \n",
       "4        +  TTATATGCTCAGACATGTCT        0.35         0.533894  \n",
       "5        +  ATACAAGGTCAGGCATGCCC        0.55         0.533894  \n",
       "6        +  CGACATGTATAGACTTGTTC        0.40         0.533894  \n",
       "7        +  GTACTTGTCCAAGCATGTCC        0.50         0.533894  \n",
       "8        -  AGACTTGTTTGAGCATGTCC        0.45         0.533894  \n",
       "9        -  GGCCATGCTGGGACATGTCC        0.65         0.533894  \n",
       "10       +  AGGCAAGCCTGAGCATGCCC        0.65         0.533894  \n",
       "11       -  GGGCATGCTCAGGCTTGCCT        0.65         0.533894  \n",
       "12       -  AGACTTGACCGGACTTGCTT        0.50         0.533894  \n",
       "13       +  GATCATGCCAGGACCTGCCT        0.60         0.533894  \n",
       "14       +  CCTCATGCCCGGGCCTGTCC        0.75         0.533894  \n",
       "15       +  AAGCACGCCGAGGCATGCCT        0.65         0.533894  \n",
       "16       +  GGACATGCTCAAACAAGTCT        0.45         0.533894  \n",
       "17       -  ATACATGCCGTGGCTTGCTT        0.50         0.533894  \n",
       "18       +  GAACATGCAGAGACCTGTCC        0.55         0.533894  \n",
       "19       -  GGACATGCTTGGACAAGTAC        0.50         0.533894  \n",
       "20       -  CCACATGTTCAAACTTGTCC        0.45         0.533894  \n",
       "21       +  AAACATGTACAAACATGCCA        0.35         0.533894  \n",
       "22       -  ACACATGTTTACGCATGTCC        0.45         0.533894  \n",
       "23       -  GTAGATGTACAGACATGTCT        0.40         0.533894  \n",
       "24       +  TTATATGTCTTGGCATGTCC        0.40         0.533894  \n",
       "25       +  AGACATGCCCGGACAAGTGT        0.55         0.533894  \n",
       "26       +  ATGCATGTTAAGGCATGTCT        0.40         0.533894  \n",
       "27       -  CGACATGTCCGGGCATGGTT        0.60         0.533894  \n",
       "28       +  AGGCATGCCCGGGCTCGTCC        0.75         0.533894  \n",
       "29       +  AAAGATGCCCGTACATGCCT        0.50         0.533894  \n",
       "..     ...                   ...         ...              ...  \n",
       "178      -  ATACCTGTCCGGACTTGTTT        0.45         0.533894  \n",
       "179      +  GAACTTGTCCGAACTTGCTC        0.50         0.533894  \n",
       "180      +  GGAAATGTCCAGACCTGTCT        0.50         0.533894  \n",
       "181      +  GGACATGTCGGCACATGTCC        0.60         0.533894  \n",
       "182      -  GAACTTGTCGTGACAAGTCC        0.50         0.533894  \n",
       "183      -  CCGCATGCTCGGACATGCCA        0.65         0.533894  \n",
       "184      -  GCAGATGCCAAGACATGCCT        0.55         0.533894  \n",
       "185      -  GGACACGCCTAAACATGTCC        0.55         0.533894  \n",
       "186      -  TGACATGTTCACACTTGTCT        0.40         0.533894  \n",
       "187      -  GAACAAGCTCATGCTTGCCC        0.55         0.533894  \n",
       "188      -  ACACATGCGGGGACATGCCT        0.60         0.533894  \n",
       "189      -  GGTTATGTCCAGGCATGCTC        0.55         0.533894  \n",
       "190      +  AAACTCGCCCGGACATGTTT        0.50         0.533894  \n",
       "191      -  AGCCATGCTGAGACATGCTC        0.55         0.533894  \n",
       "192      -  CAACATGCATGGGCATGCAC        0.55         0.533894  \n",
       "193      -  AATAATGTCCAGACATGTTT        0.30         0.533894  \n",
       "194      -  ATACATGCCTAGGCCTGCCC        0.60         0.533894  \n",
       "195      -  AATCATGCCCGAGCACGTCC        0.60         0.533894  \n",
       "196      -  GAACTTGCACAAACATGTCT        0.40         0.533894  \n",
       "197      +  GAGCATGTCTGTGCTTGCCC        0.60         0.533894  \n",
       "198      +  ATACAAGCGCTGGCATGCCC        0.60         0.533894  \n",
       "199      -  ATGGATGTTCAGGCATGCCC        0.55         0.533894  \n",
       "200      -  GCGTATGTCGGGACATGTCC        0.60         0.533894  \n",
       "201      -  GGATATGTTGAGACATGCTC        0.45         0.533894  \n",
       "202      +  ATGCATGCACTGACATGTTC        0.45         0.533894  \n",
       "203      -  GGGCAAGCACAGACATGCTC        0.60         0.533894  \n",
       "204      -  CAATATGTCCAGACCTGTCT        0.45         0.533894  \n",
       "205      +  CGACCTGTCGGGACATGCCA        0.65         0.533894  \n",
       "206      +  GAATATGTCCTGGCATGTCT        0.45         0.533894  \n",
       "207      +  CTACATGCTCACACATGTCC        0.50         0.533894  \n",
       "\n",
       "[208 rows x 10 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def identifier(row):\n",
    "    return str(row['sequence_name']) + ':' + str(row['start']) + '-' + str(row['stop']) + '_' + str(row['strand'] + ';' + str(row['motif_id']))\n",
    "\n",
    "def percent_gc(row):\n",
    "    r = pd.Series(row['matched_sequence'])\n",
    "    z = r.str.count('G') + r.str.count('C')\n",
    "    x = z/len(row['matched_sequence'])\n",
    "    return x\n",
    "\n",
    "motif_dir = '/scratch/Users/tajo5912/rbg/levandowski2021/temp/sim_fimo_out/TP53_M06704_1'\n",
    "motif_name = motif_dir.split('/')[-1]\n",
    "df = pd.read_csv('%s/fimo.tsv' % (motif_dir), sep ='\\t')\n",
    "print(\"Post-processing FIMO output for %s\" % motif_name)\n",
    "df.drop(df.tail(3).index,inplace=True)\n",
    "\n",
    "df = df.sort_values('sequence_name').reset_index()\n",
    "df['identifier'] = df.apply(lambda row: identifier(row), axis=1)\n",
    "df['percent_gc'] = df.apply(lambda row: percent_gc(row), axis=1)\n",
    "df['mean_percent_gc'] = df['percent_gc'].mean()\n",
    "df = df[['sequence_name', 'start', 'stop', 'identifier', 'motif_id','score', 'strand', \n",
    "         'matched_sequence', 'percent_gc', 'mean_percent_gc']]\n",
    "#df=df.convert_dtypes()\n",
    "df['start'] = df['start'].astype(int) \n",
    "df['stop'] = df['stop'].astype(int) \n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def fimotobed(outdir):\n",
    "#     def identifier(row):\n",
    "#         return str(row['sequence_name']) + ':' + str(row['start']) + '-' + str(row['stop']) + '_' + str(row['strand'] + ';' + str(row['motif_id']))\n",
    "\n",
    "#     def percent_gc(row):\n",
    "#         r = pd.Series(row['matched_sequence'])\n",
    "#         z = r.str.count('G') + r.str.count('C')\n",
    "#         x = z/len(row['matched_sequence'])\n",
    "#         return x\n",
    "\n",
    "#     motif_dirs = glob.glob(outdir + '/temp/sim_fimo_out/*')\n",
    "#     for motif_dir in motif_dirs:\n",
    "#         if path.isdir(motif_dir):\n",
    "#             motif_name = motif_dir.split('/')[-1]\n",
    "#             df = pd.read_csv('%s/fimo.tsv' % (motif_dir), sep ='\\t')\n",
    "#             print(\"Post-processing FIMO output for %s\" % motif_name)\n",
    "#             df.drop(df.tail(3).index,inplace=True)\n",
    "#             if df.empty == True:\n",
    "#                 print('Skipping ' + motif_name + ' -no motif hits.')\n",
    "#             else:\n",
    "#                 df = df.sort_values('sequence_name').reset_index()\n",
    "#                 df['identifier'] = df.apply(lambda row: identifier(row), axis=1)\n",
    "#                 df['percent_gc'] = df.apply(lambda row: percent_gc(row), axis=1)\n",
    "#                 df['mean_percent_gc'] = df['percent_gc'].mean()\n",
    "#                 df=df.convert_dtypes()\n",
    "#                 df = df[['sequence_name', 'start', 'stop', 'identifier', 'score', 'strand', \n",
    "#                          'matched_sequence', 'percent_gc', 'mean_percent_gc']]\n",
    "#                 df.to_csv(outdir + '/motifs/simulated/' + motif_name + '.sorted.bed', sep='\\t', header=None, index=False)\n",
    "\n",
    "#fimotobed('/scratch/Users/tajo5912/rbg/test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def expt_bed(self):\n",
    "    #getting the sequence number\n",
    "    annotation_file = pd.read_csv(self.annotation, header=None)\n",
    "    sequence_num = len(annotation_file)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    #creating the locations for the all chromosomes except the last\n",
    "    dbed = {}\n",
    "    for i in range(0,(seq_per_chrom)):\n",
    "        dbed['r_' + str(i)] = [int(self.window-1)+i*(2*self.window+21), int(self.window+1)+i*(2*self.window+21)]\n",
    "\n",
    "    dfbed = pd.DataFrame.from_dict(dbed, orient='index')\n",
    "    dfbed.columns = ['start', 'stop']\n",
    "    #duplicating the dataframe for all chromosomes except the last\n",
    "    dfbed = pd.concat([dfbed]*(self.chrom_num-1))\n",
    "    #adding the chromosome name to the genomic location\n",
    "    l_expt=[]\n",
    "    for i in range(int(seq_per_chrom)):\n",
    "        for c in range(int(self.chrom_num-1)):\n",
    "            l_expt.append(str('expt_' + str(c+1)))\n",
    "    l_expt = sorted(l_expt)\n",
    "    dfbed['expt'] = l_expt\n",
    "\n",
    "    \n",
    "    dbed = {}\n",
    "    for i in range(0,(seq_per_chrom)):\n",
    "        dbed['r_' + str(i)] = [int(self.window-1)+i*(2*self.window+21), int(self.window+1)+i*(2*self.window+21)]\n",
    "\n",
    "    dfbed = pd.DataFrame.from_dict(dbed, orient='index')\n",
    "    dfbed.columns = ['start', 'stop']\n",
    "    #duplicating the dataframe for all chromosomes except the last\n",
    "    dfbed = pd.concat([dfbed]*(self.chrom_num-1))\n",
    "    #adding the chromosome name to the genomic location\n",
    "    l_sim=[]\n",
    "    for i in range(int(seq_per_chrom)):\n",
    "        for c in range(int(self.chrom_num-1)):\n",
    "            l_sim.append(str('sim_' + str(c+1)))\n",
    "    l_sim = sorted(l_sim)\n",
    "    dfbed['sim'] = l_sim\n",
    "    \n",
    "    \n",
    "    \n",
    "    #creating the locations for the last chromosome\n",
    "    dlast = {}\n",
    "    for i in range(0,(seq_last_chrom)):\n",
    "        dlast['l_' + str(i)] = [int(self.window-1)+i*(2*self.window+21), int(self.window+1)+i*(2*self.window+21)]\n",
    "    dflast = pd.DataFrame.from_dict(dlast, orient='index')\n",
    "    dflast.columns = ['start', 'stop']\n",
    "    #adding the chromosome name to the genomic location\n",
    "    l_last = []\n",
    "    for i in range(int(seq_last_chrom)):\n",
    "        l_last.append(str('expt_' + str(self.chrom_num)))\n",
    "    dflast['expt'] = l_last\n",
    "\n",
    "    #combining the bed locations\n",
    "    df = pd.concat([dfbed, dflast])\n",
    "    df = df[['expt', 'start', 'stop']]\n",
    "    \n",
    "    #creating the locations for the last chromosome\n",
    "    dlast = {}\n",
    "    for i in range(0,(seq_last_chrom)):\n",
    "        dlast['l_' + str(i)] = [int(self.window-1)+i*(2*self.window+21), int(self.window+1)+i*(2*self.window+21)]\n",
    "    dflast = pd.DataFrame.from_dict(dlast, orient='index')\n",
    "    dflast.columns = ['start', 'stop']\n",
    "    #adding the chromosome name to the genomic location\n",
    "    l_last = []\n",
    "    for i in range(int(seq_last_chrom)):\n",
    "        l_last.append(str('sim_' + str(self.chrom_num)))\n",
    "    dflast['sim'] = l_last\n",
    "\n",
    "    #combining the bed locations\n",
    "    df = pd.concat([dfbed, dflast])\n",
    "    df = df[['sim', 'start', 'stop']]\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    #saving the new annotation\n",
    "    df.to_csv(self.outdir + \"/annotations/\" + str(self.sample) + \"_experimental_genome_centered.bed\", header=None, index=False, sep='\\t')\n",
    "    ##saving the new annotation\n",
    "    df.to_csv(self.outdir + '/annotations/' + self.sample + '_experimental_genome_self.window.bed', sep='\\t', columns=[\"expt\",\"start\",\"stop\"], header = False, index = False)    \n",
    "    return print('Experimental Bed Files Genereated: ' + self.outdir + \"/annotations/\" + str(self.sample) + \"_experimental_genome_centered.bed and \" + str(self.sample) + \"_experimental_genome_self.window.bed\")\n",
    "\n",
    "\n",
    "\n",
    "def generate_bed(self):\n",
    "    if self.experimental_fimo == True:\n",
    "        #getting the sequence number\n",
    "        annotation_file = pd.read_csv(self.annotation, header=None)\n",
    "        sequence_num = len(annotation_file)\n",
    "    else:\n",
    "        sequence_num = self.sequence_num      \n",
    "\n",
    "    #creating the locations for the all chromosomes except the last\n",
    "    dbed = {}\n",
    "    for i in range(0,(seq_per_chrom)):\n",
    "        dbed['r_' + str(i)] = [int(self.window-1)+i*(2*self.window+21), int(self.window+1)+i*(2*self.window+21)]\n",
    "\n",
    "    dfbed = pd.DataFrame.from_dict(dbed, orient='index')\n",
    "    dfbed.columns = ['start', 'stop']\n",
    "    #duplicating the dataframe for all chromosomes except the last\n",
    "    dfbed = pd.concat([dfbed]*(self.chrom_num-1))\n",
    "    #adding the chromosome name to the genomic location\n",
    "    l_sim=[]\n",
    "    for i in range(int(seq_per_chrom)):\n",
    "        for c in range(int(self.chrom_num-1)):\n",
    "            l_sim.append(str('sim_' + str(c+1)))\n",
    "    l_sim = sorted(l_sim)\n",
    "    dfbed['sim'] = l_sim\n",
    "\n",
    "    #creating the locations for the last chromosome\n",
    "    dlast = {}\n",
    "    for i in range(0,(seq_last_chrom)):\n",
    "        dlast['l_' + str(i)] = [int(self.window-1)+i*(2*self.window+21), int(self.window+1)+i*(2*self.window+21)]\n",
    "    dflast = pd.DataFrame.from_dict(dlast, orient='index')\n",
    "    dflast.columns = ['start', 'stop']\n",
    "    #adding the chromosome name to the genomic location\n",
    "    l_last = []\n",
    "    for i in range(int(seq_last_chrom)):\n",
    "        l_last.append(str('sim_' + str(self.chrom_num)))\n",
    "    dflast['sim'] = l_last\n",
    "\n",
    "    #combining the bed locations\n",
    "    df = pd.concat([dfbed, dflast])\n",
    "    df = df[['sim', 'start', 'stop']]\n",
    "    #saving the new annotation\n",
    "    df.to_csv(self.outdir + \"/annotations/\" + str(self.sample) + \"_simulated_centered.bed\", header=None, index=False, sep='\\t')\n",
    "\n",
    "\n",
    "\n",
    "    ##saving the new annotation\n",
    "    df.to_csv(self.outdir + '/annotations/' + self.sample + '_simulated_window.bed', sep='\\t', columns=[\"sim\",\"start\",\"stop\"], header = False, index = False)\n",
    "    return print('Simulated Bed Files Genereated: ' + self.outdir + \"/annotations/\" + str(self.sample) + \"_simulated_centered.bed and \" + str(self.sample) + \"_simulated_self.window.bed\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
